<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ATACseqQC_workshop2020 â€¢ workshop2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ATACseqQC_workshop2020">
<meta property="og:description" content="workshop2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">workshop2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ATACseqQC.html">ATACseqQC_workshop2020</a>
    </li>
    <li>
      <a href="../articles/ChIPpeakAnno.html">ChIPpeakAnno_workshop2020</a>
    </li>
    <li>
      <a href="../articles/motifStack.html">motifStack_workshop2020</a>
    </li>
    <li>
      <a href="../articles/trackViewer.html">trackViewer_workshop2020</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/workshop2020">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ATACseqQC_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ATACseqQC_workshop2020</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/workshop2020/blob/master/vignettes/ATACseqQC.Rmd"><code>vignettes/ATACseqQC.Rmd</code></a></small>
      <div class="hidden name"><code>ATACseqQC.Rmd</code></div>

    </div>

    
    
<div id="prepare-for-inputs" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-for-inputs" class="anchor"></a>Prepare for inputs</h2>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"jianhong/workshop2020"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">## input the sample bamFile from the installed packaged</span>
<span class="co">## the data is pre-processed chr1 data of GSM1155959</span>
<span class="no">extfilePath</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ATACseqQC"</span>,
                           <span class="kw">package</span><span class="kw">=</span><span class="st">"workshop2020"</span>, <span class="kw">mustWork</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">extfilePath</span>)</pre></body></html></div>
<pre><code>## [1] "GL3.chr1.bam"           "GL3.chr1.bam.bai"       "GL3.chr1.rmdup.bam"    
## [4] "GL3.chr1.rmdup.bam.bai"</code></pre>
</div>
<div id="estimate-the-library-complexity" class="section level2">
<h2 class="hasAnchor">
<a href="#estimate-the-library-complexity" class="anchor"></a>Estimate the library complexity</h2>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">## load the library</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ATACseqQC</span>)</pre></body></html></div>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## No methods found in package 'IRanges' for request: 'values' when loading 'GenomicScores'</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co">## mapping status, such as mapping rate, </span>
<span class="co">## duplicate rate, genome-wide distribution, </span>
<span class="co">## mapping quality, and contamination.</span>
<span class="co">## NOTE: requires sorted BAM files with duplicate reads marked as input.</span>
<span class="co">## library complexity</span>
<span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/estimateLibComplexity.html">estimateLibComplexity</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/readsDupFreq.html">readsDupFreq</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">extfilePath</span>, <span class="st">"GL3.chr1.bam"</span>)))</pre></body></html></div>
<p><img src="ATACseqQC_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="figure">
<img src="../../../../tmp/RtmpyIF4I8/temp_libpath55c630fbc8e/workshop2020/vignettes/libComp.jpg" alt="library complexity curve" width="500px"><p class="caption">
library complexity curve
</p>
</div>
</div>
<div id="fragment-size-distribution" class="section level2">
<h2 class="hasAnchor">
<a href="#fragment-size-distribution" class="anchor"></a>Fragment size distribution</h2>
<p>First, there should be a large proportion of reads with less than 100 bp, which represents the nucleosome-free region. Second, the fragment size distribution should have a clear periodicity, which is evident in the inset figure, indicative of nucleosome occupacy (present in integer multiples).</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="co">## set bam file name, replace file.path(extfilePath, "GL3.chr1.rmdup.bam")</span>
<span class="co">## by your bam file name</span>
<span class="no">bamfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">extfilePath</span>, <span class="st">"GL3.chr1.rmdup.bam"</span>)
(<span class="no">bamfile.labels</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".rmdup.bam"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="no">bamfile</span>)))</pre></body></html></div>
<pre><code>## [1] "GL3.chr1"</code></pre>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="co">## generate fragment size distribution</span>
<span class="no">fragSize</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/fragSizeDist.html">fragSizeDist</a></span>(<span class="no">bamfile</span>, <span class="no">bamfile.labels</span>)</pre></body></html></div>
<p><img src="ATACseqQC_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div id="adjust-the-read-start-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#adjust-the-read-start-sites" class="anchor"></a>Adjust the read start sites</h2>
<p>Tn5 transposase has been shown to bind as a dimer and inserts two adaptors into accessible DNA locations separated by 9 bp.</p>
<p>Therefore, for downstream analysis, such as peak-calling and footprinting, all reads in input bamfile need to be shifted. The function <code>shiftGAlignmentsList</code> can be used to shift the reads. By default, all reads aligning to the positive strand are offset by +4bp, and all reads aligning to the negative strand are offset by -5bp.</p>
<p>The adjusted reads will be written into a new bamfile for peak calling or footprinting.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">## bamfile tags to be read in</span>
<span class="no">possibleTag</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html">combn</a></span>(<span class="no">LETTERS</span>, <span class="fl">2</span>)
<span class="no">possibleTag</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">possibleTag</span>[<span class="fl">1</span>, ], <span class="no">possibleTag</span>[<span class="fl">2</span>, ]),
                 <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">possibleTag</span>[<span class="fl">2</span>, ], <span class="no">possibleTag</span>[<span class="fl">1</span>, ]))
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Rsamtools</span>)</pre></body></html></div>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: 'Biostrings'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     strsplit</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">bamTop100</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html">scanBam</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html">BamFile</a></span>(<span class="no">bamfile</span>, <span class="kw">yieldSize</span> <span class="kw">=</span> <span class="fl">100</span>),
                     <span class="kw">param</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">ScanBamParam</a></span>(<span class="kw">tag</span><span class="kw">=</span><span class="no">possibleTag</span>))<span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">tag</span>
<span class="no">tags</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">bamTop100</span>)[<span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span>(<span class="no">bamTop100</span>)<span class="kw">==</span><span class="fl">100</span>]
<span class="no">tags</span></pre></body></html></div>
<pre><code>##  [1] "AS" "MD" "PG" "XG" "NM" "XM" "XN" "XO" "XS" "YS" "YT"</code></pre>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="co">## files will be output into outPath</span>
<span class="no">outPath</span> <span class="kw">&lt;-</span> <span class="st">"splited"</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="no">outPath</span>)
<span class="co">## shift the coordinates of 5'ends of alignments in the bam file</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg38</span>)</pre></body></html></div>
<pre><code>## Loading required package: BSgenome</code></pre>
<pre><code>## Loading required package: rtracklayer</code></pre>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">seqlev</span> <span class="kw">&lt;-</span> <span class="st">"chr1"</span> <span class="co">## subsample data for quick run</span>
<span class="no">which</span> <span class="kw">&lt;-</span> <span class="fu">as</span>(<span class="fu">seqinfo</span>(<span class="no">Hsapiens</span>)[<span class="no">seqlev</span>], <span class="st">"GRanges"</span>)
<span class="no">gal</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/readBamFile.html">readBamFile</a></span>(<span class="no">bamfile</span>, <span class="kw">tag</span><span class="kw">=</span><span class="no">tags</span>, <span class="kw">which</span><span class="kw">=</span><span class="no">which</span>, <span class="kw">asMates</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">bigFile</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">shiftedBamfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outPath</span>, <span class="st">"shifted.bam"</span>)
<span class="no">gal1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/shiftGAlignmentsList.html">shiftGAlignmentsList</a></span>(<span class="no">gal</span>, <span class="kw">outbam</span><span class="kw">=</span><span class="no">shiftedBamfile</span>)</pre></body></html></div>
</div>
<div id="promotertranscript-body-pt-score" class="section level2">
<h2 class="hasAnchor">
<a href="#promotertranscript-body-pt-score" class="anchor"></a>Promoter/Transcript body (PT) score</h2>
<p>PT score is calculated as the coverage of promoter divided by the coverage of its transcript body. PT score will show if the signal is enriched in promoters.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg38.knownGene</span>)</pre></body></html></div>
<pre><code>## Loading required package: GenomicFeatures</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">txs</span> <span class="kw">&lt;-</span> <span class="fu">transcripts</span>(<span class="no">TxDb.Hsapiens.UCSC.hg38.knownGene</span>)
<span class="no">pt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/PTscore.html">PTscore</a></span>(<span class="no">gal1</span>, <span class="no">txs</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">pt</span>$<span class="no">log2meanCoverage</span>, <span class="no">pt</span>$<span class="no">PT_score</span>, <span class="kw">pch</span><span class="kw">=</span><span class="fl">16</span>, <span class="kw">cex</span> <span class="kw">=</span> <span class="fl">.5</span>,
     <span class="kw">xlab</span><span class="kw">=</span><span class="st">"log2 mean coverage"</span>,
     <span class="kw">ylab</span><span class="kw">=</span><span class="st">"Promoter vs Transcript"</span>); <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="kw">h</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"red"</span>)</pre></body></html></div>
<p><img src="ATACseqQC_files/figure-html/unnamed-chunk-6-1.png" width="700"> ## Transcription Start Site (TSS) Enrichment Score</p>
<p>TSS enrichment score is a raio between aggregate distribution of reads centered on TSSs and that flanking the corresponding TSSs. TSS score = the depth of TSS (1000 bp each side) / the depth of end flanks (100bp each end). TSS enrichment score is calculated according to the definition at <a href="https://www.encodeproject.org/data-standards/terms/#enrichment">https://www.encodeproject.org/data-standards/terms/#enrichment</a>. Transcription start site (TSS) enrichment values are dependent on the reference files used; cutoff values for high quality data are listed in the following table from <a href="https://www.encodeproject.org/atac-seq/">https://www.encodeproject.org/atac-seq/</a>.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">tsse</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/TSSEscore.html">TSSEscore</a></span>(<span class="no">gal1</span>, <span class="no">txs</span>)
<span class="no">log2TSSE</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="no">tsse</span>$<span class="no">TSS.enrichment.score</span>)
<span class="no">log2TSSE</span> <span class="kw">&lt;-</span> <span class="no">log2TSSE</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">log2TSSE</span>)]
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span>(<span class="no">log2TSSE</span>),
     <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">"log2TSSE"</span>, <span class="kw">main</span> <span class="kw">=</span> <span class="st">""</span>); <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="kw">v</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"red"</span>)</pre></body></html></div>
<p><img src="ATACseqQC_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div id="split-reads" class="section level2">
<h2 class="hasAnchor">
<a href="#split-reads" class="anchor"></a>Split reads</h2>
<p>The shifted reads will be split into different bins, namely nucleosome free, mononucleosome, dinucleosome, and trinucleosome. Shifted reads that do not fit into any of the above bins will be discarded. Splitting reads is a time-consuming step because we are using random forest to classify the fragments based on fragment length, GC content and conservation scores.</p>
<p>By default, we assign the top 10% of short reads (reads below 100_bp) as nucleosome-free regions and the top 10% of intermediate length reads as (reads between 180 and 247 bp) mononucleosome. This serves as the training set to classify the rest of the fragments using random forest. The number of the tree will be set to 2 times of square root of the length of the training set.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="co">## run program for chromosome 1 only</span>
<span class="no">txs</span> <span class="kw">&lt;-</span> <span class="no">txs</span>[<span class="fu">seqnames</span>(<span class="no">txs</span>) <span class="kw">%in%</span> <span class="st">"chr1"</span>]
<span class="no">genome</span> <span class="kw">&lt;-</span> <span class="no">Hsapiens</span>
<span class="co">## split the reads into NucleosomeFree, mononucleosome, </span>
<span class="co">## dinucleosome and trinucleosome.</span>
<span class="co">## and save the binned alignments into bam files.</span>
<span class="no">objs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/splitGAlignmentsByCut.html">splitGAlignmentsByCut</a></span>(<span class="no">gal1</span>, <span class="kw">txs</span><span class="kw">=</span><span class="no">txs</span>, <span class="kw">genome</span><span class="kw">=</span><span class="no">genome</span>,
                              <span class="kw">outPath</span> <span class="kw">=</span> <span class="no">outPath</span>)
<span class="co">## list the files generated by splitGAlignmentsByCut.</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">outPath</span>)</pre></body></html></div>
<pre><code>##  [1] "dinucleosome.bam"       "dinucleosome.bam.bai"   "inter1.bam"            
##  [4] "inter1.bam.bai"         "inter2.bam"             "inter2.bam.bai"        
##  [7] "inter3.bam"             "inter3.bam.bai"         "mononucleosome.bam"    
## [10] "mononucleosome.bam.bai" "NucleosomeFree.bam"     "NucleosomeFree.bam.bai"
## [13] "others.bam"             "others.bam.bai"         "shifted.bam"           
## [16] "shifted.bam.bai"        "trinucleosome.bam"      "trinucleosome.bam.bai"</code></pre>
</div>
<div id="heatmap-and-coverage-curve-for-nucleosome-positions" class="section level2">
<h2 class="hasAnchor">
<a href="#heatmap-and-coverage-curve-for-nucleosome-positions" class="anchor"></a>Heatmap and coverage curve for nucleosome positions</h2>
<p>By averaging the signal across all active TSSs, we should observe that nucleosome-free fragments are enriched at the TSSs, whereas the nucleosome-bound fragments should be enriched both upstream and downstream of the active TSSs and display characteristic phasing of upstream and downstream nucleosomes. Because ATAC-seq reads are concentrated at regions of open chromatin, users should see a strong nucleosome signal at the +1 nucleosome, but the signal decreases at the +2, +3 and +4 nucleosomes.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ChIPpeakAnno</span>)
<span class="no">bamfiles</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outPath</span>,
                     <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"NucleosomeFree.bam"</span>,
                     <span class="st">"mononucleosome.bam"</span>,
                     <span class="st">"dinucleosome.bam"</span>,
                     <span class="st">"trinucleosome.bam"</span>))
<span class="no">TSS</span> <span class="kw">&lt;-</span> <span class="fu">promoters</span>(<span class="no">txs</span>, <span class="kw">upstream</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">TSS</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">TSS</span>)
<span class="co">## estimate the library size for normalization</span>
(<span class="no">librarySize</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/estLibSize.html">estLibSize</a></span>(<span class="no">bamfiles</span>))</pre></body></html></div>
<pre><code>## splited/NucleosomeFree.bam splited/mononucleosome.bam 
##                      10140                       3426 
##   splited/dinucleosome.bam  splited/trinucleosome.bam 
##                       2748                          0</code></pre>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="co">## calculate the signals around TSSs.</span>
<span class="no">NTILE</span> <span class="kw">&lt;-</span> <span class="fl">101</span>
<span class="no">dws</span> <span class="kw">&lt;-</span> <span class="no">ups</span> <span class="kw">&lt;-</span> <span class="fl">1010</span>
<span class="no">sigs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/enrichedFragments.html">enrichedFragments</a></span>(<span class="kw">gal</span><span class="kw">=</span><span class="no">objs</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"NucleosomeFree"</span>,
                                     <span class="st">"mononucleosome"</span>,
                                     <span class="st">"dinucleosome"</span>,
                                     <span class="st">"trinucleosome"</span>)],
                          <span class="kw">TSS</span><span class="kw">=</span><span class="no">TSS</span>,
                          <span class="kw">librarySize</span><span class="kw">=</span><span class="no">librarySize</span>,
                          <span class="kw">seqlev</span><span class="kw">=</span><span class="no">seqlev</span>,
                          <span class="kw">TSS.filter</span><span class="kw">=</span><span class="fl">0.5</span>,
                          <span class="kw">n.tile</span> <span class="kw">=</span> <span class="no">NTILE</span>,
                          <span class="kw">upstream</span> <span class="kw">=</span> <span class="no">ups</span>,
                          <span class="kw">downstream</span> <span class="kw">=</span> <span class="no">dws</span>)
<span class="co">## log2 transformed signals</span>
<span class="no">sigs.log2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">sigs</span>, <span class="kw">function</span>(<span class="no">.ele</span>) <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="no">.ele</span>+<span class="fl">1</span>))
<span class="co">#plot heatmap</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedHeatmap.html">featureAlignedHeatmap</a></span>(<span class="no">sigs.log2</span>, <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html">reCenterPeaks</a></span>(<span class="no">TSS</span>, <span class="kw">width</span><span class="kw">=</span><span class="no">ups</span>+<span class="no">dws</span>),
                      <span class="kw">zeroAt</span><span class="kw">=</span><span class="fl">.5</span>, <span class="kw">n.tile</span><span class="kw">=</span><span class="no">NTILE</span>)</pre></body></html></div>
<p><img src="ATACseqQC_files/figure-html/unnamed-chunk-9-1.png" width="384"></p>
Here is the heatmap for the full dataset.
<div class="figure">
<img src="../../../../tmp/RtmpyIF4I8/temp_libpath55c630fbc8e/workshop2020/vignettes/heatmap.jpg" alt="metagene" width="500px"><p class="caption">
metagene
</p>
</div>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="co">## get signals normalized for nucleosome-free and nucleosome-bound regions.</span>
<span class="no">out</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedDistribution.html">featureAlignedDistribution</a></span>(<span class="no">sigs</span>,
                                  <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html">reCenterPeaks</a></span>(<span class="no">TSS</span>, <span class="kw">width</span><span class="kw">=</span><span class="no">ups</span>+<span class="no">dws</span>),
                                  <span class="kw">zeroAt</span><span class="kw">=</span><span class="fl">.5</span>, <span class="kw">n.tile</span><span class="kw">=</span><span class="no">NTILE</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"l"</span>,
                                  <span class="kw">ylab</span><span class="kw">=</span><span class="st">"Averaged coverage"</span>)</pre></body></html></div>
<pre><code>## Warning in featureAlignedDistribution(sigs, reCenterPeaks(TSS, width = ups + :
## cvglists contain NA values. NA value will be omit.</code></pre>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="co">## rescale the nucleosome-free and nucleosome signals to 0~1</span>
<span class="no">range01</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>){(<span class="no">x</span>-<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">x</span>))/(<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">x</span>)-<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">x</span>))}
<span class="no">out</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">out</span>, <span class="fl">2</span>, <span class="no">range01</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html">matplot</a></span>(<span class="no">out</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"l"</span>, <span class="kw">xaxt</span><span class="kw">=</span><span class="st">"n"</span>,
        <span class="kw">xlab</span><span class="kw">=</span><span class="st">"Position (bp)"</span>,
        <span class="kw">ylab</span><span class="kw">=</span><span class="st">"Fraction of signal"</span>);<span class="fu"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span>(<span class="fl">1</span>, <span class="kw">at</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">0</span>, <span class="fl">100</span>, <span class="kw">by</span><span class="kw">=</span><span class="fl">10</span>)+<span class="fl">1</span>,
     <span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"-1K"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(-<span class="fl">800</span>, <span class="fl">800</span>, <span class="kw">by</span><span class="kw">=</span><span class="fl">200</span>), <span class="st">"1K"</span>),
     <span class="kw">las</span><span class="kw">=</span><span class="fl">2</span>);<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="kw">v</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">0</span>, <span class="fl">100</span>, <span class="kw">by</span><span class="kw">=</span><span class="fl">10</span>)+<span class="fl">1</span>, <span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"gray"</span>)</pre></body></html></div>
<p><img src="ATACseqQC_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div id="plot-footprints" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-footprints" class="anchor"></a>plot Footprints</h2>
<p>ATAC-seq footprints infer factor occupancy genome-wide. The <code>factorFootprints</code> function uses <code>matchPWM</code> to predict the binding sites using the input position weight matrix (PWM). Then it calculates and plots the accumulated coverage for those binding sites to show the status of the occupancy genome-wide. Unlike CENTIPEDE, the footprints generated here do not take the conservation (PhyloP) into consideration. <code>factorFootprints</code> function could also accept the binding sites as a GRanges object.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="co">## foot prints</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MotifDb</span>)</pre></body></html></div>
<pre><code>## See system.file("LICENSE", package="MotifDb") for use restrictions.</code></pre>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="no">CTCF</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(<span class="no">MotifDb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CTCF"</span>))
<span class="no">CTCF</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>(<span class="no">CTCF</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">CTCF</span><span class="kw">[[</span><span class="fl">1</span>]], <span class="kw">digits</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>##      1    2    3     4     5     6     7     8     9    10    11    12    13
## A 0.10 0.16 0.30 0.072 0.012 0.786 0.024 0.122 0.914 0.012 0.376 0.022 0.028
## C 0.36 0.21 0.10 0.826 0.966 0.024 0.620 0.494 0.010 0.008 0.010 0.022 0.002
## G 0.12 0.41 0.44 0.050 0.012 0.108 0.336 0.056 0.048 0.976 0.602 0.606 0.962
## T 0.42 0.22 0.16 0.052 0.010 0.082 0.020 0.328 0.028 0.004 0.012 0.350 0.008
##      14    15    16    17   18   19
## A 0.024 0.096 0.424 0.086 0.12 0.34
## C 0.016 0.818 0.024 0.532 0.35 0.26
## G 0.880 0.038 0.522 0.326 0.12 0.32
## T 0.080 0.048 0.030 0.056 0.41 0.08</code></pre>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="no">sigs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/factorFootprints.html">factorFootprints</a></span>(<span class="no">shiftedBamfile</span>, <span class="kw">pfm</span><span class="kw">=</span><span class="no">CTCF</span><span class="kw">[[</span><span class="fl">1</span>]],
                         <span class="kw">genome</span><span class="kw">=</span><span class="no">genome</span>,
                         <span class="kw">min.score</span><span class="kw">=</span><span class="st">"90%"</span>, <span class="kw">seqlev</span><span class="kw">=</span><span class="no">seqlev</span>,
                         <span class="kw">upstream</span><span class="kw">=</span><span class="fl">100</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">100</span>)</pre></body></html></div>
<img src="ATACseqQC_files/figure-html/unnamed-chunk-13-1.png" width="700"> Here is the CTCF footprints for the full dataset.
<div class="figure">
<img src="../../../../tmp/RtmpyIF4I8/temp_libpath55c630fbc8e/workshop2020/vignettes/CtCFfootprints.png" alt="CTCF footprints" width="500px"><p class="caption">
CTCF footprints
</p>
</div>
</div>
<div id="v-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#v-plot" class="anchor"></a>V-plot</h2>
<p>V-plot is a plot to visualize fragment midpoint vs length for a given transcription factors.</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="no">vp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ATACseqQC/man/vPlot.html">vPlot</a></span>(<span class="no">shiftedBamfile</span>, <span class="kw">pfm</span><span class="kw">=</span><span class="no">CTCF</span><span class="kw">[[</span><span class="fl">1</span>]],
            <span class="kw">genome</span><span class="kw">=</span><span class="no">genome</span>, <span class="kw">min.score</span><span class="kw">=</span><span class="st">"90%"</span>, <span class="kw">seqlev</span><span class="kw">=</span><span class="no">seqlev</span>,
            <span class="kw">upstream</span><span class="kw">=</span><span class="fl">200</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">200</span>,
            <span class="kw">ylim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">30</span>, <span class="fl">250</span>), <span class="kw">bandwidth</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">1</span>))</pre></body></html></div>
<img src="ATACseqQC_files/figure-html/unnamed-chunk-15-1.png" width="700"> Here is the CTCF vPlot for the full dataset.
<div class="figure">
<img src="../../../../tmp/RtmpyIF4I8/temp_libpath55c630fbc8e/workshop2020/vignettes/CTCFvPlot.png" alt="CTCF vPlot" width="500px"><p class="caption">
CTCF vPlot
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
