<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>motifStack_workshop2020 • workshop2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="motifStack_workshop2020">
<meta property="og:description" content="workshop2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">workshop2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ATACseqQC.html">ATACseqQC_workshop2020</a>
    </li>
    <li>
      <a href="../articles/ChIPpeakAnno.html">ChIPpeakAnno_workshop2020</a>
    </li>
    <li>
      <a href="../articles/motifStack.html">motifStack_workshop2020</a>
    </li>
    <li>
      <a href="../articles/trackViewer.html">trackViewer_workshop2020</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/workshop2020">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="motifStack_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="motifStack_files/kePrint-0.0.1/kePrint.js"></script><script src="motifStack_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="motifStack_files/d3-4.4.0/d3.v4.min.js"></script><link href="motifStack_files/plotWebLogo-0.0.1/d3-plotWebLogo.css" rel="stylesheet">
<script src="motifStack_files/plotWebLogo-0.0.1/d3-plotWebLogo.js"></script><script src="motifStack_files/saveSvgAsPng-0.0.1/saveSvgAsPng.js"></script><link href="motifStack_files/d3-tip-0.8.0/d3-tip.css" rel="stylesheet">
<script src="motifStack_files/d3-tip-0.8.0/d3-tip.js"></script><script src="motifStack_files/browseMotifs-binding-1.33.6/browseMotifs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>motifStack_workshop2020</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/workshop2020/blob/master/vignettes/motifStack.Rmd"><code>vignettes/motifStack.Rmd</code></a></small>
      <div class="hidden name"><code>motifStack.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>A sequence motif is a short sequence conserved in multiple amino acid or nucleic acid sequences with biological significance such as DNA binding sites, where transcription factors (TFs) bind and catalytic sites in enzymes. Motifs are often represented as position count matrix (PCM)(Table @ref(tab:motifStackPCM)), position frequency matrix (PFM)(Table @ref(tab:motifStackPFM)) or position weight matrix (PWM) (reviewed in<span class="citation"><sup>1–3</sup></span>). Sequence logos, based on information theory or probability theory, have been used as a graphical representation of sequence motifs to depict the frequencies of bases or amino acids at each conserved position<span class="citation"><sup>4–8</sup></span>.</p>
<table class="table table">
<caption>
Sample of PCM. Each row represents the counts of A, C, G and T, while each column is the counts for that nucleotide position.
</caption>
<tbody>
<tr>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
23
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
C
</td>
<td style="text-align:right;">
19
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
19
</td>
<td style="text-align:right;">
15
</td>
</tr>
<tr>
<td style="text-align:left;">
T
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table>
<table class="table table">
<caption>
Sample of PFM. Each row represents the frequency of A, C, G and T, while each column is the frequency for that nucleotide position.
</caption>
<tbody>
<tr>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
0.04
</td>
<td style="text-align:right;">
0.04
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0.96
</td>
<td style="text-align:right;">
0.04
</td>
<td style="text-align:right;">
0.09
</td>
</tr>
<tr>
<td style="text-align:left;">
C
</td>
<td style="text-align:right;">
0.83
</td>
<td style="text-align:right;">
0.70
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:right;">
0.09
</td>
<td style="text-align:right;">
0.17
</td>
</tr>
<tr>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
0.09
</td>
<td style="text-align:right;">
0.17
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.04
</td>
<td style="text-align:right;">
0.83
</td>
<td style="text-align:right;">
0.65
</td>
</tr>
<tr>
<td style="text-align:left;">
T
</td>
<td style="text-align:right;">
0.04
</td>
<td style="text-align:right;">
0.09
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:right;">
0.04
</td>
<td style="text-align:right;">
0.09
</td>
</tr>
</tbody>
</table>
<p>High-throughput experimental approaches have been developed to identify binding sequences for a large number of TFs in several organisms. The experimental methods include protein binding microarray (PBM)<span class="citation"><sup>9</sup></span>, bacterial one-hybrid (B1H)<span class="citation"><sup>10</sup></span>, systematic evolution of ligands by exponential enrichment (SELEX)<span class="citation"><sup>11,12</sup></span>, High-Throughput (HT) SELEX<span class="citation"><sup>13</sup></span>, and selective microfluidics-based ligand enrichment followed by sequencing (SMiLE-seq)<span class="citation"><sup>14</sup></span>. Several algorithms have been developed to discover motifs from the binding sequences, identified by the above experimental methods. For example, MEME Suite<span class="citation"><sup>15</sup></span> has been widely used to construct motifs for B1H and SELEX data, while Seed-and-Wobble (SW)<span class="citation"><sup>16</sup></span>, Dream5<span class="citation"><sup>17</sup></span>, and BEEML<span class="citation"><sup>18</sup></span> have been used to generate motifs for PBM data.</p>
<p>Visualization tools have been developed for representing a sequence motif as individual sequence logo, such as weblogo<span class="citation"><sup>7</sup></span>and seqlogo<span class="citation"><sup>19</sup></span>. To facilitate the comparison of multiple motifs, several motif alignment and clustering tools have been developed such as Tomtom<span class="citation"><sup>20</sup></span>, MatAlign<span class="citation"><sup>21</sup></span>, and STAMP<span class="citation"><sup>22</sup></span>. However, these tools are lacking flexibility in dispalying the similarities or differences among motifs.</p>
<p>The <em>motifStack</em> package<span class="citation"><sup>23</sup></span> is designed for for the graphic representation of multiple motifs with their similarities or distances depicted. It provides the flexibility for users to customize the clustering and graphic parameters such as distance cutoff, the font family, colors, symbols, layout styles, and ways to highlight different regions or branches.</p>
<p>In this workshop, we will demonstrate the features and flexibility of <em>motifStack</em> for visualizing the alignment of multiple motifs in various styles. In addition, we will illustrate the utility of <em>motifStack</em> for providing insights into families of related motifs using several large collections of homeodomain (HD) DNA binding motifs from human, mouse and fly. Furthermore, we will show how <em>motifStack</em> facilitated the discovery that changing the computational method used to build motifs from the same binding data can lead to artificial segregation of motifs for identical TFs using the same motif alignment method.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><em>BiocManager</em> is used to install the released version of <em>motifStack</em>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"motifStack"</span>)</pre></body></html></div>
<p>To install the development version of <em>motifStack</em>, please try</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"jianhong/motifStack"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">"motifStack"</span>)</pre></body></html></div>
<pre><code>## [1] '1.33.6'</code></pre>
<p>Please note that starting from version 1.28.0, <em>motifStack</em> requires cario (&gt;=1.6) instead of ghostscript which is for a lower version of <em>motifStack</em>. The <code>capabilities</code> function could be used to check capablilities of cario for current R session. If <em>motifStack</em> version number is smaller than 1.28.0, the full path to the executable ghostscript must be available for R session. Starting from version 1.33.2, <em>motifStack</em> does not require cario or ghostscript any more. It will use cario if cario (&gt;=1.6) is installed or use ghostscript if <code>gs</code> command is available. Otherwise, <em>motifStack</em> will use embed font to plot the sequence logo.</p>
<p>In this tutorial, we will illustrate the functionalities of <em>motifStack</em> available in version &gt;=1.33.3.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">"motifStack"</span>)<span class="kw">&lt;</span><span class="st">"1.33.3"</span>){
  <span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"jianhong/motifStack"</span>, <span class="kw">upgrade</span><span class="kw">=</span><span class="st">"never"</span>,
                       <span class="kw">build_vignettes</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">force</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
}</pre></body></html></div>
</div>
<div id="inputs-of-motifstack" class="section level2">
<h2 class="hasAnchor">
<a href="#inputs-of-motifstack" class="anchor"></a>Inputs of <em>motifStack</em>
</h2>
<p>The inputs of <em>motifStack</em> are one or more motifs represented as PCM or PFM. Motifs in PFM format are available in most of the motif databases such as TRANSFAC<span class="citation"><sup>24</sup></span>, JASPAR<span class="citation"><sup>25</sup></span>, CIS-BP<span class="citation"><sup>26</sup></span> and HOCOMOCO<span class="citation"><sup>27</sup></span>. More databases could be found at <a href="http://meme-suite.org/db/motifs" class="uri">http://meme-suite.org/db/motifs</a>.</p>
<p>In addition, <code>MotifDb</code> package contains a collection of DNA-binding sequence motifs in PFM format. 9933 PFMs were collected from 11 sources: cisbp_1.02<span class="citation"><sup>26</sup></span>, FlyFactorSurvey<span class="citation"><sup>28</sup></span>, HOCOMOCO<span class="citation"><sup>27</sup></span>, HOMER<span class="citation"><sup>29</sup></span>, hPDI<span class="citation"><sup>30</sup></span>, JASPAR<span class="citation"><sup>25</sup></span>, Jolma2013<span class="citation"><sup>31</sup></span>, ScerTF<span class="citation"><sup>32</sup></span>, stamlab<span class="citation"><sup>33</sup></span>, SwissRegulon<span class="citation"><sup>34</sup></span> and UniPROBE<span class="citation"><sup>9</sup></span>.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MotifDb</span>)
(<span class="no">pfm</span> <span class="kw">&lt;-</span> <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(<span class="no">MotifDb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"hsapiens"</span>, <span class="st">"CTCF"</span>)))</pre></body></html></div>
<pre><code>## MotifDb object of length 12
## | Created from downloaded public sources: 2013-Aug-30
## | 12 position frequency matrices from 8 sources:
## |        HOCOMOCOv10:    2
## | HOCOMOCOv11-core-A:    2
## |        JASPAR_2014:    1
## |        JASPAR_CORE:    1
## |         jaspar2016:    1
## |         jaspar2018:    2
## |          jolma2013:    1
## |       SwissRegulon:    2
## | 1 organism/s
## |           Hsapiens:   12
## Hsapiens-HOCOMOCOv10-CTCFL_HUMAN.H10MO.A 
## Hsapiens-HOCOMOCOv10-CTCF_HUMAN.H10MO.A 
## Hsapiens-HOCOMOCOv11-core-A-CTCFL_HUMAN.H11MO.0.A 
## Hsapiens-HOCOMOCOv11-core-A-CTCF_HUMAN.H11MO.0.A 
## Hsapiens-JASPAR_CORE-CTCF-MA0139.1 
## ...
## Hsapiens-jaspar2018-CTCF-MA0139.1 
## Hsapiens-jaspar2018-CTCFL-MA1102.1 
## Hsapiens-jolma2013-CTCF 
## Hsapiens-SwissRegulon-CTCFL.SwissRegulon 
## Hsapiens-SwissRegulon-CTCF.SwissRegulon</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">pfm</span><span class="kw">[[</span><span class="st">"Hsapiens-jolma2013-CTCF"</span>]]</pre></body></html></div>
<pre><code>##            1          2          3           4            5           6
## A 0.45533165 0.17369230 0.00000000 0.005067300 0.0623177134 0.004530681
## C 0.04584383 0.02478455 0.94098399 0.000000000 0.9324324324 0.995469319
## G 0.14550798 0.75930256 0.00000000 0.988123515 0.0007777562 0.000000000
## T 0.35331654 0.04222059 0.05901601 0.006809184 0.0044720980 0.000000000
##            7           8            9          10         11          12
## A 0.50758368 0.007861919 0.0000000000 0.002594670 0.76366002 0.007318263
## C 0.45336471 0.640791614 0.9998872985 0.003436185 0.02095981 0.131403480
## G 0.01804393 0.000000000 0.0001127015 0.001612903 0.11744724 0.799154334
## T 0.02100767 0.351346467 0.0000000000 0.992356241 0.09793293 0.062123923
##           13           14         15          16        17
## A 0.01376768 0.0002326934 0.03978832 0.151956584 0.3644148
## C 0.38797773 0.0000000000 0.01117209 0.281976578 0.3105268
## G 0.00000000 0.9856893543 0.73137985 0.008169095 0.2124384
## T 0.59825459 0.0140779523 0.21765974 0.557897744 0.1126200</code></pre>
<p>The <code>plotMotifLogo</code> function can be used to plot sequence logo for motifs in PFM format (Figure @ref(fig:motifStackplotMotifLogo)). As default, the nucleotides in a sequence logo will be drawn proportional to their information content (IC)<span class="citation"><sup>8</sup></span>.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">motifStack</span>)
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/plotMotifLogo.html">plotMotifLogo</a></span>(<span class="no">pfm</span><span class="kw">[[</span><span class="st">"Hsapiens-jolma2013-CTCF"</span>]])</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotMotifLogo-1.png" alt="Plot sequence logo for motif in PFM format." width="576"><p class="caption">
Plot sequence logo for motif in PFM format.
</p>
</div>
<p>A PFM can be formated as an <code>pfm</code> object which is the preferred format for <em>motifStack</em> (Figure @ref(fig:motifStackplotMotifPFM)).</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">matrix.bin</span> <span class="kw">&lt;-</span> <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(<span class="no">MotifDb</span>, <span class="st">"bin_SANGER"</span>)
<span class="no">motif</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"pfm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="no">matrix.bin</span><span class="kw">[[</span><span class="fl">1</span>]], <span class="kw">name</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">matrix.bin</span>)[<span class="fl">1</span>])
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifStack.html">motifStack</a></span>(<span class="no">motif</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotMotifPFM-1.png" alt="Plot sequence logo for a `pfm` object." width="576"><p class="caption">
Plot sequence logo for a <code>pfm</code> object.
</p>
</div>
<p>Similary, a PCM can be converted to a <code>pcm</code> object easily (Figure @ref(fig:motifStackplotMotifPCM)).</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"motifStack"</span>)
(<span class="no">pcm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"bin_SOLEXA.pcm"</span>)))</pre></body></html></div>
<pre><code>##   V1 V2  V3   V4   V5   V6   V7  V8   V9
## 1  A  | 462    0 1068 1025 1068   0 1019
## 2  C  |  71   60    0   24    0 993    0
## 3  G  | 504    0    0    0    0  12    0
## 4  T  |  31 1008    0   19    0  63   49</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">pcm</span> <span class="kw">&lt;-</span> <span class="no">pcm</span>[,<span class="fl">3</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">pcm</span>)]
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">pcm</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"A"</span>,<span class="st">"C"</span>,<span class="st">"G"</span>,<span class="st">"T"</span>) <span class="co"># must have rownames</span>
<span class="no">motif</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"pcm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">pcm</span>), <span class="kw">name</span><span class="kw">=</span><span class="st">"bin_SOLEXA"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/pcm-class.html">plot</a></span>(<span class="no">motif</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotMotifPCM-1.png" alt="Plot sequence logo for a `pcm` object." width="576"><p class="caption">
Plot sequence logo for a <code>pcm</code> object.
</p>
</div>
<p>In addition, <code>importMatrix</code> function can be used to import the motifs into a <code>pcm</code> or <code>pfm</code> object from files exported by TRANSFAC<span class="citation"><sup>24</sup></span>, JASPAR<span class="citation"><sup>25</sup></span> or CIS-BP<span class="citation"><sup>26</sup></span>.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">motifs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/importMatrix.html">importMatrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">path</span>, <span class="st">"*.pcm"</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                       <span class="kw">format</span> <span class="kw">=</span> <span class="st">"pcm"</span>, <span class="kw">to</span> <span class="kw">=</span> <span class="st">"pcm"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifStack.html">motifStack</a></span>(<span class="no">motifs</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackimportMatrix-1.png" alt="Plot sequence logo for data imported by `importMatrix`" width="576"><p class="caption">
Plot sequence logo for data imported by <code>importMatrix</code>
</p>
</div>
</div>
<div id="plot-single-sequence-logo" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-single-sequence-logo" class="anchor"></a>Plot single sequence logo</h2>
<p><em>motifStack</em> can be used to plot DNA/RNA sequence logo, affinity logo and amino acid sequence logo. It provides the flexibility for users to customize the graphic parameters such as the font family, symbol colors and markers. Different from most of the sequence logo plotting tools, <em>motifStack</em> does not need to pre-define the path for plotting. It depends on grImport/2<span class="citation"><sup>35</sup></span> package to import fonts into a path, which provides the power of <em>motifStack</em> more flexibility.</p>
</div>
<div id="plot-a-dnarnaamino-acid-sequence-logo" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-a-dnarnaamino-acid-sequence-logo" class="anchor"></a>Plot a DNA/RNA/amino acid sequence logo</h2>
<p>To plot an RNA sequence logo, simply change the rowname of the matrix from “T” to “U” (Figure @ref(fig:motifStackplotRNAmotif)).</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">rna</span> <span class="kw">&lt;-</span> <span class="no">pcm</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">rna</span>)[<span class="fl">4</span>] <span class="kw">&lt;-</span> <span class="st">"U"</span>
<span class="no">RNA_motif</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"pcm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">rna</span>), <span class="kw">name</span><span class="kw">=</span><span class="st">"RNA_motif"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/pcm-class.html">plot</a></span>(<span class="no">RNA_motif</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotRNAmotif-1.png" alt="Plot an RNA sequence logo." width="576"><p class="caption">
Plot an RNA sequence logo.
</p>
</div>
<p>If the PFM of amino acid sequence motif is provided, <em>motifStack</em> can be used to draw amino acid sequence logo (Figure @ref(fig:motifStackplotAAmotif)).</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Biostrings</span>)
<span class="no">protein</span><span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"motifStack"</span>, <span class="st">"irg.txt"</span>,
                                <span class="kw">package</span> <span class="kw">=</span> <span class="st">"workshop2020"</span>))
<span class="no">protein</span><span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">protein</span>[,<span class="fl">2</span>:<span class="fl">21</span>])
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">protein</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">AA_STANDARD</span>)
<span class="no">protein_motif</span><span class="kw">&lt;-</span><span class="fu">new</span>(<span class="st">"pcm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="no">protein</span>, <span class="kw">name</span><span class="kw">=</span><span class="st">"IRG"</span>,
                   <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/colorset.html">colorset</a></span>(<span class="kw">alphabet</span><span class="kw">=</span><span class="st">"AA"</span>,<span class="kw">colorScheme</span><span class="kw">=</span><span class="st">"chemistry"</span>),
                   <span class="kw">alphabet</span> <span class="kw">=</span> <span class="st">"AA"</span>,
                   <span class="kw">markers</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu">new</span>(<span class="st">"marker"</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"rect"</span>,
                                    <span class="kw">start</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">27</span>), <span class="kw">stop</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">13</span>,<span class="fl">49</span>),
                                    <span class="kw">gp</span><span class="kw">=</span><span class="fu">gpar</span>(<span class="kw">col</span><span class="kw">=</span><span class="st">"#009E73"</span>, <span class="kw">fill</span><span class="kw">=</span><span class="fl">NA</span>, <span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>))))
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/pcm-class.html">plot</a></span>(<span class="no">protein_motif</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotAAmotif-1.png" alt="Plot amino acid sequence logo." width="576"><p class="caption">
Plot amino acid sequence logo.
</p>
</div>
<p><em>motifStack</em> can also be used to plot other types of logos (Figure @ref(fig:motifStackplotANYlogo)).</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">m</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">0</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">10</span>,
            <span class="kw">dimnames</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="st">"motifStack"</span>, <span class="st">""</span>)<span class="kw">[[</span><span class="fl">1</span>]],
                            <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="st">"motifStack"</span>, <span class="st">""</span>)<span class="kw">[[</span><span class="fl">1</span>]]))
<span class="kw">for</span>(<span class="no">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq.int</a></span>(<span class="fl">10</span>)) <span class="no">m</span>[<span class="no">i</span>, <span class="no">i</span>] <span class="kw">&lt;-</span> <span class="fl">1</span>
<span class="no">ms</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"pfm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="no">m</span>, <span class="kw">name</span><span class="kw">=</span><span class="st">"motifStack"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/pcm-class.html">plot</a></span>(<span class="no">ms</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotANYlogo-1.png" alt="Plot a user-defined logo." width="576"><p class="caption">
Plot a user-defined logo.
</p>
</div>
<p>The sequence logo can be visulized with different fonts and colors (Figure @ref(fig:motifStackchangeFont)).</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">motif</span>$<span class="no">color</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/colorset.html">colorset</a></span>(<span class="kw">colorScheme</span> <span class="kw">=</span> <span class="st">'blindnessSafe'</span>)
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/pcm-class.html">plot</a></span>(<span class="no">motif</span>, <span class="kw">font</span><span class="kw">=</span><span class="st">"mono,Courier"</span>, <span class="kw">fontface</span><span class="kw">=</span><span class="st">"plain"</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackchangeFont-1.png" alt="Plot sequence logo with different fonts and colors." width="576"><p class="caption">
Plot sequence logo with different fonts and colors.
</p>
</div>
<div id="plot-sequence-logo-with-specified-regions-highlighted-using-the-markers-slot" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-sequence-logo-with-specified-regions-highlighted-using-the-markers-slot" class="anchor"></a>Plot sequence logo with specified regions highlighted using the markers slot</h3>
<p>The <code>markers</code> slot of a <code>pfm</code> or <code>pcm</code> object can be assigned by a list of <code>marker</code> objects. There are three type of markers, “rect”, “line” and “text”. If <code>markers</code> slot is assigned, markers can be plotted (Figure @ref(fig:motifStackwithMarkers)).</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">markerRect</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"marker"</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"rect"</span>, <span class="kw">start</span><span class="kw">=</span><span class="fl">6</span>, <span class="kw">stop</span><span class="kw">=</span><span class="fl">7</span>,
                  <span class="kw">gp</span><span class="kw">=</span><span class="fu">gpar</span>(<span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">fill</span><span class="kw">=</span><span class="fl">NA</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"orange"</span>, <span class="kw">lwd</span><span class="kw">=</span><span class="fl">3</span>))
<span class="no">markerRect2</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"marker"</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"rect"</span>, <span class="kw">start</span><span class="kw">=</span><span class="fl">2</span>,
                   <span class="kw">gp</span><span class="kw">=</span><span class="fu">gpar</span>(<span class="kw">lty</span><span class="kw">=</span><span class="fl">3</span>, <span class="kw">fill</span><span class="kw">=</span><span class="fl">NA</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"green"</span>, <span class="kw">lwd</span><span class="kw">=</span><span class="fl">2</span>))
<span class="no">markerLine</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"marker"</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"line"</span>, <span class="kw">start</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">stop</span><span class="kw">=</span><span class="fl">7</span>,
                  <span class="kw">gp</span><span class="kw">=</span><span class="fu">gpar</span>(<span class="kw">lwd</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"darkblue"</span>))
<span class="no">markerText</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"marker"</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"text"</span>, <span class="kw">start</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">5</span>),
                  <span class="kw">label</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"*"</span>, <span class="st">"core"</span>), <span class="kw">gp</span><span class="kw">=</span><span class="fu">gpar</span>(<span class="kw">cex</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"red"</span>))
<span class="no">motif</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"pcm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">pcm</span>), <span class="kw">name</span><span class="kw">=</span><span class="st">"bin_SOLEXA"</span>,
             <span class="kw">markers</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">markerRect</span>, <span class="no">markerLine</span>, <span class="no">markerText</span>, <span class="no">markerRect2</span>))
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/pcm-class.html">plot</a></span>(<span class="no">motif</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackwithMarkers-1.png" alt="Plot a sequence logo with selected regions highlighted." width="576"><p class="caption">
Plot a sequence logo with selected regions highlighted.
</p>
</div>
</div>
</div>
<div id="plot-motifs-with-ggplot2" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-motifs-with-ggplot2" class="anchor"></a>Plot motifs with ggplot2</h2>
<p>Sequence logo can also be plotted by <code>geom_motif</code> function which is compatible with <em>ggplot2</em><span class="citation"><sup>36</sup></span> (Figure @ref(fig:motifStackggplot2)).</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">len</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">motifs</span>)
<span class="no">motifs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">markers</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">markerRect2</span>, <span class="no">markerRect</span>)
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="fl">.5</span>, <span class="kw">y</span><span class="kw">=</span>(<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq.int</a></span>(<span class="no">len</span>)-<span class="fl">.5</span>)/<span class="no">len</span>,
                 <span class="kw">width</span><span class="kw">=</span><span class="fl">.75</span>, <span class="kw">height</span><span class="kw">=</span><span class="fl">1</span>/(<span class="no">len</span>+<span class="fl">1</span>))
<span class="no">df</span>$<span class="no">motif</span> <span class="kw">&lt;-</span> <span class="no">motifs</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">x</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">y</span>, <span class="kw">width</span><span class="kw">=</span><span class="no">width</span>, <span class="kw">height</span><span class="kw">=</span><span class="no">height</span>, <span class="kw">motif</span><span class="kw">=</span><span class="no">motif</span>)) +
  <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/geom_motif.html">geom_motif</a></span>(<span class="kw">use.xy</span> <span class="kw">=</span> <span class="fl">TRUE</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="fl">0</span>, <span class="fl">1</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span>(<span class="fl">0</span>, <span class="fl">1</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackggplot2-1.png" alt="Plot sequence logo with ggplot2." width="576"><p class="caption">
Plot sequence logo with ggplot2.
</p>
</div>
</div>
<div id="plot-multiple-sequence-logos" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-multiple-sequence-logos" class="anchor"></a>Plot multiple sequence logos</h2>
<p><em>motifStack</em> is designed to visulize the alignment of multiple motifs and facilitate the analysis of binding site diversity/conservation within families of TFs and evolution of binding sites among different species. There are many functions in <em>motifStack</em> such as <code>plotMotifLogoStack</code>, <code>motifCloud</code>, <code>plotMotifStackWithRadialPhylog</code>, <code>motifPiles</code>, and <code>motifCircos</code> for plotting motifs in different layouts, <code>motifSignature</code> for generating motif signatures by merging similar motifs. To make it easy to use, we created one workflow function called <code>motifStack</code>, which can be used to draw a single DNA, RNA or amino acid motif as a sequence logo, as well as to align motifs, generate motif signatures and plot aligned motifs as a stack, phylogeny tree, radial tree or cloud of sequence logos.</p>
<p>To align motifs, <code>DNAmotifAlignment</code> function implements Ungapped Smith–Waterman (local) alignment based on column comparison scores calculated by Kullback-Leibler distance<span class="citation"><sup>37</sup></span>. <code>DNAmotifAlignment</code> requires the inputs be sorted by the distance of motifs. To get the sorted motifs for alignment, distances can be calculated using motifDistance in <code>motIV</code> (R implementation of STAMP<span class="citation"><sup>22</sup></span>). Distances generated from MatAlign<span class="citation"><sup>21</sup></span> are also supported with a helper function <code><a href="https://rdrr.io/pkg/ade4/man/newick2phylog.html">ade4::newick2phylog</a></code><span class="citation"><sup>38</sup></span>.</p>
<div id="plot-sequence-logo-stack" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-sequence-logo-stack" class="anchor"></a>Plot sequence logo stack</h3>
<p>As shown in above, <code>motifStack</code> function can be used to plot motifs as a stack of sequence logos easily. By setting the layout to “tree” (Figure @ref(fig:motifStacktreeLayout)) or “phylog” (Figure @ref(fig:motifStackphylogLayout)), <code>motifStack</code> will plot the stack of sequence logos with a linear tree.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifStack.html">motifStack</a></span>(<span class="no">motifs</span>, <span class="kw">layout</span> <span class="kw">=</span> <span class="st">"tree"</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStacktreeLayout-1.png" alt="Plot sequence logo stack with tree layout." width="576"><p class="caption">
Plot sequence logo stack with tree layout.
</p>
</div>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifStack.html">motifStack</a></span>(<span class="no">motifs</span>, <span class="kw">layout</span><span class="kw">=</span><span class="st">"phylog"</span>, <span class="kw">f.phylog</span><span class="kw">=</span><span class="fl">.15</span>, <span class="kw">f.logo</span><span class="kw">=</span><span class="fl">0.25</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackphylogLayout-1.png" alt="Plot sequence logo stack with phylog layout." width="576"><p class="caption">
Plot sequence logo stack with phylog layout.
</p>
</div>
<p>Even though the layout of “phylog” allows more motifs to be plotted comparing to the layout of “tree”, both layouts only allow small sets of motifs to be readily compared. However, moderate throughput analysis of TFs has generated 100s to 1000s of motifs, which make the visualization of their relationships more challenging. To reduce the total number of motifs depicted, clusters of highly similar motifs can be merged with the <code>motifSignature</code> function using a distance threshold. The distance threshold is user-determined; a low threshold will merge only extremely similar motifs while a higher threshold can group motifs with lower similarity.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MotifDb</span>)
<span class="no">matrix.fly</span> <span class="kw">&lt;-</span> <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(<span class="no">MotifDb</span>, <span class="st">"FlyFactorSurvey"</span>)
<span class="no">motifs2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>(<span class="no">matrix.fly</span>)
<span class="co">## format the name</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">motifs2</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"(_[\\.0-9]+)*_FBgn\\d+$"</span>, <span class="st">""</span>,
                       <span class="fu">elementMetadata</span>(<span class="no">matrix.fly</span>)$<span class="no">providerName</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">motifs2</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"[^a-zA-Z0-9]"</span>, <span class="st">"_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">motifs2</span>))

<span class="no">motifs2</span> <span class="kw">&lt;-</span> <span class="no">motifs2</span>[<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">motifs2</span>))]
<span class="co">## subsample motifs</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="no">motifs2</span>, <span class="fl">50</span>)
<span class="co">## cluster the motifs</span>
<span class="no">hc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/clusterMotifs.html">clusterMotifs</a></span>(<span class="no">pfms</span>)
<span class="co">## convert the hclust to phylog object</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ade4</span>)
<span class="no">phylog</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ade4/man/newick2phylog.html">hclust2phylog</a></span>(<span class="no">hc</span>)
<span class="co">## reorder the pfms by the order of hclust</span>
<span class="no">leaves</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">phylog</span>$<span class="no">leaves</span>)
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="no">pfms</span>[<span class="no">leaves</span>]
<span class="co">## create a list of pfm objects</span>
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span>(<span class="no">pfms</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">pfms</span>),
               <span class="kw">FUN</span><span class="kw">=</span><span class="kw">function</span>(<span class="no">.pfm</span>, <span class="no">.name</span>){
                 <span class="fu">new</span>(<span class="st">"pfm"</span>,<span class="kw">mat</span><span class="kw">=</span><span class="no">.pfm</span>, <span class="kw">name</span><span class="kw">=</span><span class="no">.name</span>)})
<span class="co">## extract the motif signatures</span>
<span class="no">motifSig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifSignature.html">motifSignature</a></span>(<span class="no">pfms</span>, <span class="no">phylog</span>, <span class="kw">cutoffPval</span> <span class="kw">=</span> <span class="fl">0.0001</span>, <span class="kw">min.freq</span><span class="kw">=</span><span class="fl">1</span>)</pre></body></html></div>
<p>Logos for these merged motifs can be plotted on the same dendrograms, using the relative size of the logos to reflect the number of TFs that contribute to the merged motif. Compared with the original dendrogram, the number of TFs with similar motifs and the relationships between the merged motifs can be more easily compared. The 50 motifs shown in the original image are now shown as only 28 motif clusters/signatures, with the members of each cluster clearly indicated. By setting the colors of dendrogram via parameter col.tree, the colors of lable of leaves via parameter col.leaves, and the colors of annotation bar via parameter col.anno, <code>motifPiles</code> function can mark each motif with multiple annotation colors (Figure @ref(fig:motifStackmotifPiles)).</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co">## get the signatures from object of motifSignature</span>
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifSig-class.html">signatures</a></span>(<span class="no">motifSig</span>)
<span class="co">## get the group color for each signature</span>
<span class="no">gpCol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifSig-class.html">sigColor</a></span>(<span class="no">motifSig</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RColorBrewer</span>)
<span class="no">color</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="fl">12</span>, <span class="st">"Set3"</span>)
<span class="co">## plot the logo stack with pile style.</span>
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifPiles.html">motifPiles</a></span>(<span class="kw">phylog</span><span class="kw">=</span><span class="no">phylog</span>, <span class="kw">pfms</span><span class="kw">=</span><span class="no">pfms</span>, <span class="kw">pfms2</span><span class="kw">=</span><span class="no">sig</span>,
            <span class="kw">col.tree</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="no">color</span>, <span class="kw">each</span><span class="kw">=</span><span class="fl">5</span>),
            <span class="kw">col.leaves</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="no">color</span>), <span class="kw">each</span><span class="kw">=</span><span class="fl">5</span>),
            <span class="kw">col.pfms2</span><span class="kw">=</span><span class="no">gpCol</span>,
            <span class="kw">r.anno</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.02</span>, <span class="fl">0.03</span>, <span class="fl">0.04</span>),
            <span class="kw">col.anno</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">colors</a></span>(), <span class="fl">50</span>),
                          <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">colors</a></span>(), <span class="fl">50</span>),
                          <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">colors</a></span>(), <span class="fl">50</span>)),
            <span class="kw">motifScale</span><span class="kw">=</span><span class="st">"logarithmic"</span>,
            <span class="kw">plotIndex</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackmotifPiles-1.png" alt="Plot sequence logo stack by motifPiles." width="576"><p class="caption">
Plot sequence logo stack by motifPiles.
</p>
</div>
</div>
<div id="plot-sequence-logo-cloud" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-sequence-logo-cloud" class="anchor"></a>Plot sequence logo cloud</h3>
<p>To emphasize the weight of motif signatures, motif signatures can be plotted as a circular (Figure @ref(fig:motifStackmotifCloud)) or rectanglular word clouds with motif size representing the number of motifs that contributed to the signature. The larger the motif size, the larger the number of motifs within that motif signature.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="co">## draw the motifs with a word-cloud style.</span>
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifCloud.html">motifCloud</a></span>(<span class="no">motifSig</span>, <span class="kw">scale</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">6</span>, <span class="fl">.5</span>), <span class="kw">layout</span><span class="kw">=</span><span class="st">"cloud"</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackmotifCloud-1.png" alt="Plot sequence logo cloud." width="576"><p class="caption">
Plot sequence logo cloud.
</p>
</div>
</div>
<div id="plot-sequence-logo-circos" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-sequence-logo-circos" class="anchor"></a>Plot sequence logo circos</h3>
<p>The sequence logo cloud will loose the cluster information. Plotting motifs as sequence logo circos is a good solution to visualize motif signatures with dendrogram (Figure @ref(fig:motifStackplotMotifStackWithRadialPhylog)).</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="co">## plot the logo stack with radial style.</span>
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/plotMotifStackWithRadialPhylog.html">plotMotifStackWithRadialPhylog</a></span>(<span class="kw">phylog</span><span class="kw">=</span><span class="no">phylog</span>, <span class="kw">pfms</span><span class="kw">=</span><span class="no">sig</span>,
                              <span class="kw">circle</span><span class="kw">=</span><span class="fl">0.4</span>, <span class="kw">cleaves</span> <span class="kw">=</span> <span class="fl">0.3</span>,
                              <span class="kw">clabel.leaves</span> <span class="kw">=</span> <span class="fl">0.5</span>,
                              <span class="kw">col.bg</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="no">color</span>, <span class="kw">each</span><span class="kw">=</span><span class="fl">5</span>), <span class="kw">col.bg.alpha</span><span class="kw">=</span><span class="fl">0.3</span>,
                              <span class="kw">col.leaves</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="no">color</span>), <span class="kw">each</span><span class="kw">=</span><span class="fl">5</span>),
                              <span class="kw">col.inner.label.circle</span><span class="kw">=</span><span class="no">gpCol</span>,
                              <span class="kw">inner.label.circle.width</span><span class="kw">=</span><span class="fl">0.03</span>,
                              <span class="kw">angle</span><span class="kw">=</span><span class="fl">350</span>, <span class="kw">circle.motif</span><span class="kw">=</span><span class="fl">1.2</span>,
                              <span class="kw">motifScale</span><span class="kw">=</span><span class="st">"logarithmic"</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackplotMotifStackWithRadialPhylog-1.png" alt="Plot sequence logo circos." width="576"><p class="caption">
Plot sequence logo circos.
</p>
</div>
<p><code>motifCircos</code> function provide the possibility to plot the motif signatures with their original sequence logos in the circos style (Figure @ref(fig:motifStackmotifCircos)).</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co">## plot the logo stack with cirsoc style.</span>
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifCircos.html">motifCircos</a></span>(<span class="kw">phylog</span><span class="kw">=</span><span class="no">phylog</span>, <span class="kw">pfms</span><span class="kw">=</span><span class="no">pfms</span>, <span class="kw">pfms2</span><span class="kw">=</span><span class="no">sig</span>,
            <span class="kw">col.tree.bg</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="no">color</span>, <span class="kw">each</span><span class="kw">=</span><span class="fl">5</span>), <span class="kw">col.tree.bg.alpha</span><span class="kw">=</span><span class="fl">0.3</span>,
            <span class="kw">col.leaves</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="no">color</span>), <span class="kw">each</span><span class="kw">=</span><span class="fl">5</span>),
            <span class="kw">col.inner.label.circle</span><span class="kw">=</span><span class="no">gpCol</span>,
            <span class="kw">inner.label.circle.width</span><span class="kw">=</span><span class="fl">0.03</span>,
            <span class="kw">col.outer.label.circle</span><span class="kw">=</span><span class="no">gpCol</span>,
            <span class="kw">outer.label.circle.width</span><span class="kw">=</span><span class="fl">0.03</span>,
            <span class="kw">r.rings</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.02</span>, <span class="fl">0.03</span>, <span class="fl">0.04</span>),
            <span class="kw">col.rings</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">colors</a></span>(), <span class="fl">50</span>),
                           <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">colors</a></span>(), <span class="fl">50</span>),
                           <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">colors</a></span>(), <span class="fl">50</span>)),
            <span class="kw">angle</span><span class="kw">=</span><span class="fl">350</span>, <span class="kw">motifScale</span><span class="kw">=</span><span class="st">"logarithmic"</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackmotifCircos-1.png" alt="Plot sequence logo circos by motifCiros." width="576"><p class="caption">
Plot sequence logo circos by motifCiros.
</p>
</div>
</div>
</div>
<div id="plot-motifs-with-d3-js-in-a-webpage" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-motifs-with-d3-js-in-a-webpage" class="anchor"></a>Plot motifs with d3.js in a webpage</h2>
<p>Interactive plot can be generated using browseMotifs function which leverages the d3.js library. All motifs on the plot are draggable and the plot can be easily exported as a Scalable Vector Graphics (SVG) file.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/browseMotifs.html">browseMotifs</a></span>(<span class="kw">pfms</span> <span class="kw">=</span> <span class="no">pfms</span>, <span class="kw">phylog</span> <span class="kw">=</span> <span class="no">phylog</span>, <span class="kw">layout</span><span class="kw">=</span><span class="st">"radialPhylog"</span>,
             <span class="kw">yaxis</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">xaxis</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">baseWidth</span><span class="kw">=</span><span class="fl">6</span>, <span class="kw">baseHeight</span> <span class="kw">=</span> <span class="fl">15</span>)</pre></body></html></div>
<div class="figure">
<div id="htmlwidget-714d4c9e76eef7e11d19" style="width:1152px;height:1152px;" class="browseMotifs html-widget"></div>
<script type="application/json" data-for="htmlwidget-714d4c9e76eef7e11d19">{"x":{"elements":{"name":"Root","children":[{"name":"Int45","children":[{"name":"Int31","children":[{"name":"Int20","children":[{"name":"H2_0_SOLEXA","motif":[[[0.312380952380952,0.0533333333333333,0.135238095238095,0.499047619047619],[0.11047619047619,0.0247619047619048,0.0590476190476191,0.805714285714286],[0.325714285714286,0.00380952380952381,0.0380952380952381,0.632380952380952],[0.952380952380952,0.0304761904761905,0,0.0171428571428571],[0.0438095238095238,0,0.100952380952381,0.855238095238095],[0.0171428571428571,0.0742857142857143,0.188571428571429,0.72],[0.594285714285714,0.019047619047619,0.333333333333333,0.0533333333333333],[0.314285714285714,0.266666666666667,0.278095238095238,0.140952380952381]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"Int18","children":[{"name":"Int11","children":[{"name":"CG15696_SOLEXA","motif":[[[0.195652173913043,0.226449275362319,0.208333333333333,0.369565217391304],[0.134057971014493,0.186594202898551,0.101449275362319,0.577898550724638],[0.126811594202899,0.0181159420289855,0.0199275362318841,0.835144927536232],[0.641304347826087,0,0.128623188405797,0.230072463768116],[0.994565217391304,0,0.00543478260869565,0],[0,0.0036231884057971,0.0199275362318841,0.976449275362319],[0.00181159420289855,0.00181159420289855,0.0199275362318841,0.976449275362319],[0.422101449275362,0,0.576086956521739,0.00181159420289855]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"Antp_SOLEXA","motif":[[[0.208706786171575,0.256722151088348,0.206145966709347,0.32842509603073],[0.140845070422535,0.195262483994878,0.0851472471190781,0.578745198463508],[0.0678617157490397,0,0.0102432778489117,0.921895006402049],[0.793854033290653,0.0480153649167734,0.00960307298335467,0.148527528809219],[0.961587708066581,0.0384122919334187,0,0],[0.000640204865556978,0.0749039692701665,0.0371318822023047,0.887323943661972],[0.0646606914212548,0.0332906530089629,0.462227912932138,0.439820742637644],[0.580665813060179,0.0134443021766965,0.361715749039693,0.0441741357234315]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false}]},{"name":"Int15","children":[{"name":"Lbe_SOLEXA","motif":[[[0.119205298013245,0.304635761589404,0.102649006622517,0.473509933774834],[0.0463576158940397,0,0.043046357615894,0.910596026490066],[0.966887417218543,0.033112582781457,0,0],[1,0,0,0],[0,0.347682119205298,0.0827814569536424,0.56953642384106],[0.109271523178808,0.231788079470199,0.268211920529801,0.390728476821192],[0.821192052980132,0.0397350993377483,0.139072847682119,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false},{"name":"bcd_FlyReg","motif":[[[0.1875,0.375,0.0625,0.375],[0.166666666666667,0.0625,0.0208333333333333,0.75],[0.9375,0.0208333333333333,0.0208333333333333,0.0208333333333333],[0.979166666666667,0,0.0208333333333333,0],[0.0208333333333333,0,0.333333333333333,0.645833333333333],[0.0208333333333333,0.916666666666667,0,0.0625],[0.0416666666666667,0.541666666666667,0.0625,0.354166666666667],[0.104166666666667,0.25,0.375,0.270833333333333]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false}]}]}]},{"name":"Int27","children":[{"name":"Int9","children":[{"name":"Int5","children":[{"name":"CG13424_SOLEXA","motif":[[[0.191266375545852,0.297816593886463,0.196506550218341,0.314410480349345],[0.229933614966807,0.269764634882317,0.248642124321062,0.251659625829813],[0.135652173913043,0.304347826086957,0.113623188405797,0.446376811594203],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0,0,0,1],[0.184347826086957,0.0515942028985507,0.227826086956522,0.536231884057971],[0.463341346153846,0.00420673076923077,0.517427884615385,0.0150240384615385],[0.187443946188341,0.295067264573991,0.313004484304933,0.204484304932735]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"Otp_SOLEXA","motif":[[[0.123674911660777,0.34982332155477,0.0918727915194346,0.434628975265018],[0.0123674911660777,0.00353356890459364,0.00706713780918728,0.977031802120141],[0.973498233215548,0.0106007067137809,0.0088339222614841,0.00706713780918728],[0.973498233215548,0.019434628975265,0.00706713780918728,0],[0.00706713780918728,0.0106007067137809,0.0106007067137809,0.971731448763251],[0.0353356890459364,0.0441696113074205,0.106007067137809,0.814487632508834],[0.653710247349823,0.00353356890459364,0.316254416961131,0.0265017667844523]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false}]},{"name":"Int8","children":[{"name":"Dll_Cell","motif":[[[0.0434782608695652,0,0,0.956521739130435],[0.91304347826087,0,0.0869565217391304,0],[1,0,0,0],[0,0,0,1],[0.173913043478261,0,0.0869565217391304,0.739130434782609],[0.478260869565217,0,0.347826086956522,0.173913043478261],[0,0.521739130434783,0.173913043478261,0.304347826086957]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false},{"name":"Int3","children":[{"name":"Zen2_Cell","motif":[[[0.153846153846154,0.269230769230769,0.0769230769230769,0.5],[0.269230769230769,0.346153846153846,0.230769230769231,0.153846153846154],[0.115384615384615,0.269230769230769,0.192307692307692,0.423076923076923],[0.0384615384615385,0,0,0.961538461538462],[1,0,0,0],[1,0,0,0],[0,0.0384615384615385,0,0.961538461538462],[0,0,0.384615384615385,0.615384615384615],[0.846153846153846,0,0.153846153846154,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":60,"height":15,"xaxis":false,"yaxis":false},{"name":"Int2","children":[{"name":"CG32532_Cell","motif":[[[0.0869565217391304,0.260869565217391,0.0869565217391304,0.565217391304348],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0,0,0,1],[0,0,0.0869565217391304,0.91304347826087],[0.521739130434783,0,0.434782608695652,0.0434782608695652]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false},{"name":"CG13424_Cell","motif":[[[0.285714285714286,0.238095238095238,0.19047619047619,0.285714285714286],[0.0952380952380952,0.333333333333333,0.0476190476190476,0.523809523809524],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0,0,0,1],[0.142857142857143,0,0.19047619047619,0.666666666666667],[0.380952380952381,0,0.619047619047619,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false}]}]}]}]},{"name":"Int24","children":[{"name":"BH2_Cell","motif":[[[0.0952380952380952,0.333333333333333,0.333333333333333,0.238095238095238],[0.285714285714286,0.0952380952380952,0.0476190476190476,0.571428571428571],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0.238095238095238,0,0,0.761904761904762],[0.142857142857143,0.0476190476190476,0.0952380952380952,0.714285714285714],[0,0,1,0],[0.238095238095238,0.19047619047619,0.523809523809524,0.0476190476190476]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":60,"height":15,"xaxis":false,"yaxis":false},{"name":"Int10","children":[{"name":"Gsc_Cell","motif":[[[0.363636363636364,0.363636363636364,0.227272727272727,0.0454545454545455],[0.181818181818182,0.227272727272727,0.181818181818182,0.409090909090909],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0,0,0,1],[0,1,0,0],[0,0.636363636363636,0.0909090909090909,0.272727272727273]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"Int4","children":[{"name":"Bcd_Cell","motif":[[[0.0454545454545455,0.272727272727273,0.454545454545455,0.227272727272727],[0.136363636363636,0.227272727272727,0.136363636363636,0.5],[0,0,0,1],[0.909090909090909,0,0,0.0909090909090909],[1,0,0,0],[0,0,0.0454545454545455,0.954545454545455],[0,1,0,0],[0,0.954545454545455,0,0.0454545454545455]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"Oc_Cell","motif":[[[0.263157894736842,0.105263157894737,0.0526315789473684,0.578947368421053],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0,0,0.105263157894737,0.894736842105263],[0,1,0,0],[0,0.894736842105263,0.0526315789473684,0.0526315789473684]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false}]}]}]}]}]},{"name":"Int35","children":[{"name":"Mirr_Cell","motif":[[[0.439024390243902,0.146341463414634,0.146341463414634,0.268292682926829],[0.439024390243902,0.0975609756097561,0.24390243902439,0.219512195121951],[0.536585365853659,0.170731707317073,0.146341463414634,0.146341463414634],[0.48780487804878,0.024390243902439,0.024390243902439,0.463414634146341],[0.707317073170732,0,0.0487804878048781,0.24390243902439],[1,0,0,0],[0,1,0,0],[1,0,0,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"Int1","children":[{"name":"So_SOLEXA","motif":[[[0.31638418079096,0.192090395480226,0.178907721280603,0.312617702448211],[0.295668549905838,0.173258003766478,0.220338983050847,0.310734463276836],[0.322033898305085,0.163841807909605,0.284369114877589,0.229755178907721],[0.00376647834274953,0.109227871939736,0.00753295668549906,0.879472693032015],[0.0225988700564972,0,0.977401129943503,0],[1,0,0,0],[0,0.0150659133709981,0.0263653483992467,0.958568738229755],[1,0,0,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"Optix_SOLEXA","motif":[[[0.361702127659574,0.161347517730496,0.223404255319149,0.25354609929078],[0.324468085106383,0.196808510638298,0.210992907801418,0.267730496453901],[0.24645390070922,0.164893617021277,0.382978723404255,0.205673758865248],[0,0.118794326241135,0,0.881205673758865],[0,0,1,0],[1,0,0,0],[0,0,0,1],[0.893617021276596,0.076241134751773,0.0301418439716312,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false}]}]}]},{"name":"Int48","children":[{"name":"Int46","children":[{"name":"Int42","children":[{"name":"Int37","children":[{"name":"Int13","children":[{"name":"kni_SANGER","motif":[[[0.590909090909091,0.136363636363636,0.272727272727273,0],[0.608695652173913,0.0434782608695652,0.0434782608695652,0.304347826086957],[1,0,0,0],[0.91304347826087,0,0,0.0869565217391304],[0.130434782608696,0.304347826086957,0.347826086956522,0.217391304347826],[0,0.0869565217391304,0.130434782608696,0.782608695652174],[0.739130434782609,0,0.260869565217391,0],[0.0434782608695652,0,0.956521739130435,0],[0.565217391304348,0.0434782608695652,0.391304347826087,0],[0.130434782608696,0.130434782608696,0.521739130434783,0.217391304347826],[0,1,0,0],[1,0,0,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":78,"height":15,"xaxis":false,"yaxis":false},{"name":"knrl_SANGER","motif":[[[0.736842105263158,0.0526315789473684,0.0526315789473684,0.157894736842105],[0.95,0.05,0,0],[0.9,0,0.05,0.05],[0.05,0.45,0.2,0.3],[0,0.1,0,0.9],[0.7,0,0.3,0],[0,0,1,0],[0.35,0,0.55,0.1],[0,0,0.6,0.4],[0,1,0,0],[0.8,0,0.2,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false}]},{"name":"Int30","children":[{"name":"hb_NAR","motif":[[[0,0.730769230769231,0.115384615384615,0.153846153846154],[0.961538461538462,0,0.0384615384615385,0],[0.807692307692308,0.192307692307692,0,0],[1,0,0,0],[1,0,0,0],[1,0,0,0],[1,0,0,0],[0.961538461538462,0.0384615384615385,0,0],[0.461538461538462,0.384615384615385,0,0.153846153846154],[0.461538461538462,0.423076923076923,0.115384615384615,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"vnd_FlyReg","motif":[[[0.333333333333333,0,0.666666666666667,0],[0,1,0,0],[0.666666666666667,0,0,0.333333333333333],[0,1,0,0],[0,0,0,1],[0,0,0,1],[0,0.333333333333333,0.666666666666667,0],[1,0,0,0],[0.333333333333333,0,0.666666666666667,0],[0.333333333333333,0.666666666666667,0,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false}]}]},{"name":"Int40","children":[{"name":"Int33","children":[{"name":"klu_SANGER","motif":[[[0.111111111111111,0.111111111111111,0.111111111111111,0.666666666666667],[0.157894736842105,0,0.842105263157895,0],[0.05,0.5,0,0.45],[0,0,1,0],[0,0,0.35,0.65],[0.05,0,0.95,0],[0,0,1,0],[0,0,1,0],[0.1,0,0,0.9],[0,0,1,0],[0,0.1,0.55,0.35]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false},{"name":"Int21","children":[{"name":"dei_da_SANGER","motif":[[[0.764705882352941,0.235294117647059,0,0],[0.411764705882353,0.588235294117647,0,0],[0,1,0,0],[1,0,0,0],[0,0.529411764705882,0,0.470588235294118],[0.352941176470588,0.647058823529412,0,0],[0,0,0,1],[0,0,1,0],[0.0588235294117647,0,0.294117647058824,0.647058823529412],[0.294117647058824,0.235294117647059,0,0.470588235294118]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"toy_FlyReg","motif":[[[0,1,0,0],[0.2,0.8,0,0],[0.4,0.4,0.2,0],[0.2,0.8,0,0],[0,0,0,1],[0,1,0,0],[1,0,0,0],[0,0.6,0,0.4],[0,0.2,0.2,0.6],[0,0.6,0.4,0],[0.6,0,0.4,0],[0,0.2,0.2,0.6]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":78,"height":15,"xaxis":false,"yaxis":false}]}]},{"name":"Int36","children":[{"name":"sug_SANGER","motif":[[[0.0952380952380952,0.80952380952381,0,0.0952380952380952],[0.0952380952380952,0.571428571428571,0.238095238095238,0.0952380952380952],[0.142857142857143,0.238095238095238,0.619047619047619,0],[0.0952380952380952,0.476190476190476,0.0952380952380952,0.333333333333333],[0.0952380952380952,0,0.857142857142857,0.0476190476190476],[0,0,1,0],[0,0,1,0],[0,0,1,0],[0,0,1,0],[0,0,0.904761904761905,0.0952380952380952],[0,0.285714285714286,0.142857142857143,0.571428571428571],[0.19047619047619,0.666666666666667,0.0476190476190476,0.0952380952380952],[0.142857142857143,0.142857142857143,0.285714285714286,0.428571428571429],[0.0476190476190476,0.0476190476190476,0.333333333333333,0.571428571428571]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":90,"height":15,"xaxis":false,"yaxis":false},{"name":"Int23","children":[{"name":"phol_SANGER","motif":[[[0.25,0.75,0,0],[1,0,0,0],[1,0,0,0],[0.25,0.0625,0.375,0.3125],[1,0,0,0],[0,0,0,1],[0,0,1,0],[0,0,1,0],[0,1,0,0],[0,0.0625,0.9375,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"Rel_SANGER","motif":[[[0.434782608695652,0.130434782608696,0.434782608695652,0],[0.391304347826087,0,0.608695652173913,0],[0.347826086956522,0,0.565217391304348,0.0869565217391304],[0.782608695652174,0,0.217391304347826,0],[0.91304347826087,0.0434782608695652,0,0.0434782608695652],[0.652173913043478,0.0434782608695652,0,0.304347826086957],[0,0.739130434782609,0,0.260869565217391],[0.0434782608695652,0.956521739130435,0,0],[0,1,0,0],[0.130434782608696,0.869565217391304,0,0],[0,1,0,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false}]}]}]}]},{"name":"Int43","children":[{"name":"Int29","children":[{"name":"Bap_Cell","motif":[[[0.304347826086957,0.173913043478261,0.173913043478261,0.347826086956522],[0.0434782608695652,0.391304347826087,0.217391304347826,0.347826086956522],[0,0,0.0434782608695652,0.956521739130435],[0,0,0,1],[1,0,0,0],[1,0,0,0],[0,0,1,0],[0.0434782608695652,0,0,0.956521739130435],[0.304347826086957,0,0.695652173913043,0],[0.130434782608696,0.347826086956522,0.391304347826087,0.130434782608696]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"Int19","children":[{"name":"fru_PG_SANGER","motif":[[[0,0.2,0.8,0],[1,0,0,0],[1,0,0,0],[0.9,0,0,0.1],[0.8,0.2,0,0],[0.8,0,0,0.2],[0.5,0,0,0.5]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false},{"name":"slp1_NAR","motif":[[[0.195121951219512,0.0731707317073171,0.341463414634146,0.390243902439024],[0,0,0,1],[0.0731707317073171,0,0.926829268292683,0],[0,0,0,1],[0,0,0.024390243902439,0.975609756097561],[0,0,0.0487804878048781,0.951219512195122],[0.658536585365854,0,0.0975609756097561,0.24390243902439],[0.024390243902439,0.536585365853659,0.170731707317073,0.268292682926829],[0.414634146341463,0.195121951219512,0.365853658536585,0.024390243902439],[0.0975609756097561,0.317073170731707,0.0731707317073171,0.51219512195122],[0.365853658536585,0.0731707317073171,0.0975609756097561,0.463414634146341]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false}]}]},{"name":"Int41","children":[{"name":"Int38","children":[{"name":"Int12","children":[{"name":"dsf_SANGER","motif":[[[0.782608695652174,0.0434782608695652,0.0869565217391304,0.0869565217391304],[0.782608695652174,0,0.173913043478261,0.0434782608695652],[1,0,0,0],[0.956521739130435,0,0,0.0434782608695652],[0,0,1,0],[0,0,0,1],[0,0.91304347826087,0,0.0869565217391304],[0.956521739130435,0,0.0434782608695652,0],[0.478260869565217,0.173913043478261,0.130434782608696,0.217391304347826],[0.478260869565217,0.347826086956522,0,0.173913043478261]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"tll_NAR","motif":[[[0.606060606060606,0.121212121212121,0.121212121212121,0.151515151515152],[0.878787878787879,0,0.121212121212121,0],[1,0,0,0],[0.939393939393939,0.0303030303030303,0.0303030303030303,0],[0,0,1,0],[0,0.0606060606060606,0,0.939393939393939],[0,1,0,0],[1,0,0,0],[0.939393939393939,0,0.0303030303030303,0.0303030303030303],[0.515151515151515,0.303030303030303,0.0606060606060606,0.121212121212121]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false}]},{"name":"Int34","children":[{"name":"Hr4_SANGER","motif":[[[0.941176470588235,0,0.0588235294117647,0],[1,0,0,0],[0.0588235294117647,0,0.941176470588235,0],[0,0,1,0],[0,0,0,1],[0,1,0,0],[0.941176470588235,0.0588235294117647,0,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false},{"name":"disco_SANGER","motif":[[[0.526315789473684,0.0526315789473684,0.105263157894737,0.315789473684211],[0.6,0,0.2,0.2],[0.25,0,0.4,0.35],[0,0,0,1],[0.0526315789473684,0,0.947368421052632,0],[0,0,0,1],[0,1,0,0],[0.9,0,0.05,0.05],[0,1,0,0],[0.25,0.7,0.05,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false}]}]},{"name":"Int39","children":[{"name":"Int25","children":[{"name":"ttk_NAR","motif":[[[0.454545454545455,0.363636363636364,0.136363636363636,0.0454545454545455],[1,0,0,0],[0,0,1,0],[0,0,1,0],[1,0,0,0],[0,0.227272727272727,0,0.772727272727273],[0.954545454545455,0,0,0.0454545454545455],[1,0,0,0],[0.0454545454545455,0.227272727272727,0.136363636363636,0.590909090909091]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":60,"height":15,"xaxis":false,"yaxis":false},{"name":"Int22","children":[{"name":"CG12768_SANGER","motif":[[[0.347826086956522,0.0434782608695652,0,0.608695652173913],[0.260869565217391,0.260869565217391,0.173913043478261,0.304347826086957],[0,0.434782608695652,0,0.565217391304348],[0.0869565217391304,0.0434782608695652,0,0.869565217391304],[1,0,0,0],[0.0434782608695652,0.956521739130435,0,0],[0,1,0,0],[1,0,0,0],[1,0,0,0],[0.304347826086957,0.391304347826087,0,0.304347826086957],[0.434782608695652,0.391304347826087,0,0.173913043478261],[0.347826086956522,0.304347826086957,0,0.347826086956522]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":78,"height":15,"xaxis":false,"yaxis":false},{"name":"Int17","children":[{"name":"Ets21c_SANGER","motif":[[[0.631578947368421,0.0526315789473684,0.315789473684211,0],[0,0.473684210526316,0,0.526315789473684],[0.157894736842105,0,0,0.842105263157895],[0,0,0,1],[0,1,0,0],[0,1,0,0],[0,0,1,0],[0,0,0.894736842105263,0.105263157894737],[0.277777777777778,0.166666666666667,0,0.555555555555556]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":60,"height":15,"xaxis":false,"yaxis":false},{"name":"Mad_FlyReg","motif":[[[0.05,0.6,0.25,0.1],[0.05,0.25,0.35,0.35],[0.15,0,0.8,0.05],[0.15,0.4,0.4,0.05],[0,0.8,0.05,0.15],[0,0,1,0],[0.3,0.1,0.3,0.3],[0,0.75,0.2,0.05],[0.25,0,0.75,0],[0.05,0.75,0,0.2]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false}]}]}]},{"name":"Int28","children":[{"name":"wor_SANGER","motif":[[[0,1,0,0],[0.952380952380952,0,0,0.0476190476190476],[0,1,0,0],[0,1,0,0],[0,0,0,1],[0,0,1,0],[0.0526315789473684,0.631578947368421,0,0.315789473684211]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":48,"height":15,"xaxis":false,"yaxis":false},{"name":"vfl_SANGER","motif":[[[0.0555555555555556,0.0555555555555556,0.666666666666667,0.222222222222222],[0,1,0,0],[1,0,0,0],[0,0,1,0],[0,0,1,0],[0,0,0,1],[1,0,0,0],[0.166666666666667,0.0555555555555556,0.777777777777778,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false}]}]}]}]}]},{"name":"Int47","children":[{"name":"Int7","children":[{"name":"nau_SANGER","motif":[[[0.722222222222222,0,0.277777777777778,0],[0.888888888888889,0.0555555555555556,0.0555555555555556,0],[0,0.944444444444444,0,0.0555555555555556],[1,0,0,0],[0.0555555555555556,0.222222222222222,0.666666666666667,0.0555555555555556],[0,0.833333333333333,0.166666666666667,0],[0,0.0555555555555556,0,0.944444444444444],[0,0,1,0],[0.944444444444444,0,0,0.0555555555555556],[0,1,0,0],[0.166666666666667,0,0.722222222222222,0.111111111111111],[0.0714285714285714,0.857142857142857,0,0.0714285714285714]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":78,"height":15,"xaxis":false,"yaxis":false},{"name":"Fer1_SANGER","motif":[[[0.529411764705882,0.117647058823529,0.0588235294117647,0.294117647058824],[0.0588235294117647,0.647058823529412,0.117647058823529,0.176470588235294],[0.352941176470588,0.117647058823529,0.470588235294118,0.0588235294117647],[0.411764705882353,0.352941176470588,0,0.235294117647059],[0,1,0,0],[1,0,0,0],[0,0.176470588235294,0.823529411764706,0],[0,0.941176470588235,0.0588235294117647,0],[0,0,0,1],[0,0,1,0],[1,0,0,0],[0,0.941176470588235,0,0.0588235294117647],[0.0588235294117647,0,0.882352941176471,0.0588235294117647]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":84,"height":15,"xaxis":false,"yaxis":false}]},{"name":"Int44","children":[{"name":"Int16","children":[{"name":"tgo_trh_SANGER","motif":[[[0.380952380952381,0.19047619047619,0.428571428571429,0],[0,0.0476190476190476,0.0952380952380952,0.857142857142857],[1,0,0,0],[0,1,0,0],[0,0,0.952380952380952,0.0476190476190476],[0,0,0,1],[0,0,1,0],[0.904761904761905,0.0952380952380952,0,0],[0.0476190476190476,0.761904761904762,0.142857142857143,0.0476190476190476],[0.0476190476190476,0.523809523809524,0.19047619047619,0.238095238095238]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":66,"height":15,"xaxis":false,"yaxis":false},{"name":"Int14","children":[{"name":"tgo_cyc_SANGER","motif":[[[0.0526315789473684,0.105263157894737,0.842105263157895,0],[0,0,0.157894736842105,0.842105263157895],[0,1,0,0],[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1],[0,0,1,0],[0.4375,0.4375,0.125,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":60,"height":15,"xaxis":false,"yaxis":false},{"name":"Int6","children":[{"name":"Met_Clk_SANGER","motif":[[[0,0.142857142857143,0.642857142857143,0.214285714285714],[0.714285714285714,0.214285714285714,0.0714285714285714,0],[0,1,0,0],[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1],[0,0,1,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":54,"height":15,"xaxis":false,"yaxis":false},{"name":"tai_SANGER","motif":[[[0.619047619047619,0,0.142857142857143,0.238095238095238],[0.714285714285714,0.0476190476190476,0,0.238095238095238],[0.523809523809524,0.333333333333333,0.0952380952380952,0.0476190476190476],[0,1,0,0],[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0.0952380952380952,0,0.904761904761905],[0,0.0952380952380952,0.904761904761905,0],[0.333333333333333,0,0.333333333333333,0.333333333333333],[0.0952380952380952,0.761904761904762,0.0952380952380952,0.0476190476190476]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false}]}]}]},{"name":"Int32","children":[{"name":"ss_tgo_SANGER","motif":[[[0.230769230769231,0,0.769230769230769,0],[0,0,0,1],[0,1,0,0],[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,1,0,0],[0.769230769230769,0,0.230769230769231,0],[1,0,0,0],[0.153846153846154,0.153846153846154,0.0769230769230769,0.615384615384615],[0,0.384615384615385,0.615384615384615,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false},{"name":"Int26","children":[{"name":"HLHm7_SANGER","motif":[[[0,0.0555555555555556,0.888888888888889,0.0555555555555556],[0,1,0,0],[0.947368421052632,0,0.0526315789473684,0],[0,1,0,0],[0,0,1,0],[0.105263157894737,0,0,0.894736842105263],[0,0,1,0],[0,0.894736842105263,0.0526315789473684,0.0526315789473684],[0,0.947368421052632,0.0526315789473684,0],[0.894736842105263,0.0526315789473684,0.0526315789473684,0],[0.631578947368421,0.263157894736842,0.105263157894737,0]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false},{"name":"CG3065_F1_3_SANGER","motif":[[[0.238095238095238,0.0476190476190476,0.714285714285714,0],[0.0952380952380952,0.0476190476190476,0.285714285714286,0.571428571428571],[0.0476190476190476,0,0.952380952380952,0],[0,0,1,0],[0,0,1,0],[0,1,0,0],[0,0,1,0],[0,0,0.0476190476190476,0.952380952380952],[0,0,1,0],[0,0.0476190476190476,0.857142857142857,0.0952380952380952],[0,0.65,0,0.35]]],"letters":["A","C","G","T"],"color":{"A":"#00811B","C":"#2000C7","G":"#FFB32C","T":"#D00001"},"background":{"A":0.25,"C":0.25,"G":0.25,"T":0.25},"width":72,"height":15,"xaxis":false,"yaxis":false}]}]}]}]}]}]},"nodeRadius":2.5,"maxH":15,"maxW":90,"layout":"radialPhylog"},"evals":[],"jsHooks":[]}</script><p class="caption">
Plot motifs interactively.
</p>
</div>
</div>
<div id="plot-fly-data-from-different-experimental-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-fly-data-from-different-experimental-methods" class="anchor"></a>Plot fly data from different experimental methods</h2>
<p>Following scripts will show how <em>motifStack</em> facilitated the discovery that changing the experiemntal method used to build motifs from the same binding data can lead to artificial segregation of motifs for identical TFs within a motif alignment. MatAlign<span class="citation"><sup>21</sup></span> was implemented in <em>motifStack</em> by following the code from <a href="http://stormo.wustl.edu/MatAlign/" class="uri">http://stormo.wustl.edu/MatAlign/</a> to get the phylogenetic tree for given motifs. Compare to STAMP, MatAlign alignment method is superior for discriminating between the closely related motifs<span class="citation"><sup>23</sup></span>.</p>
<p>Here we clustered motifs for a smaller set of fly HDs that were determined using four different experimental methods to measure DNA binding specificity, B1H, HT-SELEX, SMiLE-seq and PBM, to confirm that the motif generation method can impact clustering. As it shown in Figre @ref(fig:motifStackFlydata) B1H motifs from different lab tend to cluster with each other (motifs 17-26, 29,30, 36-45, 50-53, 60-68, 72-100, 102-143, 159-169, 172-180, 184, 185, 187, 188, 196, 197, 202-210, 238, 239) compared to PBM data (motifs ‘eve’ in blue). PMB data (motif 227-230, 240-247) tend to have lower IC, which may be introduced by computation methods<span class="citation"><sup>23</sup></span>. HT-SELEX tend to have dimer (motif labels in red).</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">pfmpath</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">"motifStack"</span>, <span class="st">"pfms"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'workshop2020'</span>)
<span class="no">pfmfiles</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">pfmpath</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">pfmfiles</span>, <span class="no">importMatrix</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"pfm"</span>, <span class="kw">to</span><span class="kw">=</span><span class="st">"pfm"</span>)
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">pfms</span>)
<span class="co">## fix the name issue</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">pfms</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"\\."</span>, <span class="st">"_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">pfms</span>))
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">pfms</span>, <span class="kw">function</span>(<span class="no">.ele</span>){
  <span class="no">.ele</span>@<span class="kw">name</span> <span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"\\."</span>, <span class="st">"_"</span>, <span class="no">.ele</span>@<span class="kw">name</span>)
  <span class="no">.ele</span>
})
<span class="co">## get phylog</span>
<span class="no">hc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/clusterMotifs.html">clusterMotifs</a></span>(<span class="no">pfms</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ade4</span>)
<span class="no">phylog</span> <span class="kw">&lt;-</span> <span class="kw pkg">ade4</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ade4/man/newick2phylog.html">hclust2phylog</a></span>(<span class="no">hc</span>)
<span class="no">leaves</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">phylog</span>$<span class="no">leaves</span>)
<span class="no">motifs</span> <span class="kw">&lt;-</span> <span class="no">pfms</span>[<span class="no">leaves</span>]
<span class="no">motifSig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifSignature.html">motifSignature</a></span>(<span class="no">motifs</span>, <span class="no">phylog</span>,
                           <span class="kw">cutoffPval</span> <span class="kw">=</span> <span class="fl">0.0001</span>, <span class="kw">min.freq</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifSig-class.html">signatures</a></span>(<span class="no">motifSig</span>)
<span class="no">gpCol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifSig-class.html">sigColor</a></span>(<span class="no">motifSig</span>)

<span class="co">##set methods color</span>
<span class="no">colorSet</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"B1H_Mathelier"</span><span class="kw">=</span><span class="st">"#40E0D0"</span>, <span class="st">"B1H_Zhu"</span><span class="kw">=</span><span class="st">"#008080"</span>,
              <span class="st">"PBM_Busser"</span><span class="kw">=</span><span class="st">"brown"</span>, <span class="st">"PBM_Weirauch"</span><span class="kw">=</span><span class="st">"#F69156"</span>,
              <span class="st">"SELEX_Nitta"</span><span class="kw">=</span><span class="st">"blue"</span>,
              <span class="st">"SMiLE_seq_Isakova"</span><span class="kw">=</span><span class="st">"#D900D9"</span>)
<span class="no">methods</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"^.*?__"</span>, <span class="st">""</span>, <span class="no">leaves</span>))
<span class="no">lev</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="no">methods</span>)
<span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="no">methods</span>) <span class="kw">&lt;-</span> <span class="no">colorSet</span>[<span class="no">lev</span>]
<span class="no">methods</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">methods</span>)

<span class="no">leaveNames</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"__.*$"</span>, <span class="st">""</span>, <span class="no">leaves</span>)
<span class="no">leaveCol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">leaveNames</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"unc_4"</span>, <span class="st">"hbn"</span>, <span class="st">"bsh"</span>, <span class="st">"Optix"</span>,
                                     <span class="st">"al"</span>, <span class="st">"PHDP"</span>, <span class="st">"CG32532"</span>, <span class="st">"CG11294"</span>),
                   <span class="st">"#D55E00"</span>,
                   <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">leaveNames</span> <span class="kw">==</span> <span class="st">"eve"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#000000"</span>))

<span class="co">## calculate average of top 8 position information content for each motif</span>
<span class="no">icgp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">motifs</span>, <span class="no">getIC</span>), <span class="kw">function</span>(<span class="no">.ele</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">.ele</span>, <span class="kw">decreasing</span><span class="kw">=</span><span class="fl">TRUE</span>)[<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="fl">8</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">.ele</span>))]))
<span class="no">icgp.ranges</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">icgp</span>)
<span class="no">icgp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span>(<span class="no">icgp</span>, <span class="fl">10</span>, <span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#009E73"</span>, <span class="st">"#000000"</span>, <span class="st">"#D55E00"</span>))(<span class="fl">10</span>))
<span class="no">icgp.image</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/as.raster.html">as.raster</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="no">icgp</span>)), <span class="kw">ncol</span><span class="kw">=</span><span class="fl">1</span>))
<span class="no">icgp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">icgp</span>)

<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifCircos.html">motifCircos</a></span>(<span class="kw">phylog</span><span class="kw">=</span><span class="no">phylog</span>, <span class="kw">pfms</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/DNAmotifAlignment.html">DNAmotifAlignment</a></span>(<span class="no">motifs</span>),
            <span class="kw">col.tree.bg</span><span class="kw">=</span><span class="no">methods</span>, <span class="kw">col.tree.bg.alpha</span><span class="kw">=</span><span class="fl">.3</span>,
            <span class="kw">r.rings</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">.05</span>, <span class="fl">.1</span>), <span class="kw">col.rings</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">icgp</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"white"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">icgp</span>))),
            <span class="kw">col.inner.label.circle</span><span class="kw">=</span><span class="no">gpCol</span>, <span class="kw">inner.label.circle.width</span><span class="kw">=</span><span class="fl">0.03</span>,
            <span class="kw">labels.leaves</span><span class="kw">=</span><span class="no">leaveNames</span>, <span class="kw">col.leaves</span><span class="kw">=</span><span class="no">leaveCol</span>,
            <span class="kw">cleaves</span><span class="kw">=</span><span class="fl">.1</span>, <span class="kw">r.tree</span><span class="kw">=</span><span class="fl">1.3</span>, <span class="kw">r.leaves</span><span class="kw">=</span><span class="fl">.2</span>, <span class="kw">clabel.leaves</span><span class="kw">=</span><span class="fl">.7</span>,
            <span class="kw">motifScale</span><span class="kw">=</span><span class="st">"logarithmic"</span>, <span class="kw">angle</span><span class="kw">=</span><span class="fl">358</span>,
            <span class="kw">plotIndex</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">IndexCex</span><span class="kw">=</span><span class="fl">.7</span>)

<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(-<span class="fl">2.4</span>, <span class="fl">2.4</span>, <span class="kw">legend</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">colorSet</span>),
       <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/highlightCol.html">highlightCol</a></span>(<span class="no">colorSet</span>, <span class="fl">.3</span>),
       <span class="kw">border</span><span class="kw">=</span><span class="st">"white"</span>, <span class="kw">lty</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">bty</span> <span class="kw">=</span> <span class="st">"n"</span>)

<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="fl">2.05</span>, <span class="fl">2.3</span>, <span class="kw">labels</span><span class="kw">=</span><span class="st">"Information Content"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/rasterImage.html">rasterImage</a></span>(<span class="no">icgp.image</span>, <span class="fl">2</span>, <span class="fl">1.8</span>, <span class="fl">2.04</span>, <span class="fl">2.2</span>, <span class="kw">interpolate</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2.05</span>, <span class="fl">2.05</span>), <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1.8</span>, <span class="fl">2.2</span>), <span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span>(<span class="no">icgp.ranges</span>, <span class="kw">digits</span><span class="kw">=</span><span class="fl">3</span>), <span class="kw">adj</span><span class="kw">=</span><span class="fl">0</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/motifStackFlydata-1.png" alt="Fly data from different methods." width="1152"><p class="caption">
Fly data from different methods.
</p>
</div>
</div>
<div id="daglogo-visualizing-conserved-amino-acid-sequence-pattern-in-groups-based-on-probability-theory" class="section level2">
<h2 class="hasAnchor">
<a href="#daglogo-visualizing-conserved-amino-acid-sequence-pattern-in-groups-based-on-probability-theory" class="anchor"></a>dagLogo: visualizing conserved amino acid sequence pattern in groups based on probability theory</h2>
<p>As it shown in amino acid (AA) sequence logo (Figure @ref(fig:motifStackplotAAmotif)), the figure is somehow busy. To optimize the vigualization of amino acid logos, we developed the <em>dagLogo</em> package. This package can reduce the complex AA alphabets to group AAs of similar properties and provide statistical test for differential AA usage.</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="co">## load library dagLogo</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dagLogo</span>)
<span class="co">## prepare the proteome of fruitfly</span>
<span class="no">proteome</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/prepareProteomeByFTP.html">prepareProteomeByFTP</a></span>(<span class="kw">source</span> <span class="kw">=</span> <span class="st">"UniProt"</span>,
                                 <span class="kw">species</span> <span class="kw">=</span> <span class="st">"Drosophila melanogaster"</span>)
<span class="co">## read the aligned sequences</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Biostrings</span>)
<span class="no">dat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readAAStringSet</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"motifStack"</span>, <span class="st">"irg.aligned.fa"</span>,
                                   <span class="kw">package</span> <span class="kw">=</span> <span class="st">"workshop2020"</span>))
<span class="co">## prepare an object of dagPeptides Class from a Proteome object</span>
<span class="co">## set the anchor to the most conserved position </span>
<span class="no">seq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/formatSequence.html">formatSequence</a></span>(<span class="kw">seq</span> <span class="kw">=</span> <span class="no">dat</span>, <span class="kw">proteome</span> <span class="kw">=</span> <span class="no">proteome</span>,
                      <span class="kw">upstreamOffset</span> <span class="kw">=</span> <span class="fl">25</span>, <span class="kw">downstreamOffset</span> <span class="kw">=</span> <span class="fl">29</span>)
<span class="co">## build the background model </span>
<span class="no">bg_ztest</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/buildBackgroundModel.html">buildBackgroundModel</a></span>(<span class="no">seq</span>, <span class="kw">background</span> <span class="kw">=</span> <span class="st">"wholeProteome"</span>,
                                 <span class="kw">proteome</span> <span class="kw">=</span> <span class="no">proteome</span>, <span class="kw">testType</span> <span class="kw">=</span> <span class="st">"ztest"</span>)
<span class="co">## grouping based on the charge of AAs.</span>
<span class="no">t</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/testDAU.html">testDAU</a></span>(<span class="kw">dagPeptides</span> <span class="kw">=</span> <span class="no">seq</span>, <span class="kw">dagBackground</span> <span class="kw">=</span> <span class="no">bg_ztest</span>,
             <span class="kw">groupingScheme</span> <span class="kw">=</span> <span class="st">"chemistry_property_Mahler_group"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/dagLogo.html">dagLogo</a></span>(<span class="no">t</span>, <span class="kw">legend</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/dagLogo-1.png" alt="Visualizing conserved amino acid sequence pattern based on chemical properties" width="576"><p class="caption">
Visualizing conserved amino acid sequence pattern based on chemical properties
</p>
</div>
<p>The package <em>dagLogo</em> allows users to use their own grouping scheme. To make the information more informatively for this special case, we assign the color of each group by their hydrophobicity.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="co">## group of the aa by chemical properties</span>
<span class="co">## order is them by hydrophobicity</span>
<span class="no">group</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
    <span class="kw">RHK</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"K"</span>, <span class="st">"R"</span>, <span class="st">"H"</span>),
    <span class="kw">DE</span><span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"D"</span>,<span class="st">"E"</span>),
    <span class="kw">QN</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Q"</span>, <span class="st">"N"</span>),
    <span class="kw">P</span> <span class="kw">=</span> <span class="st">"P"</span>,
    <span class="kw">WYF</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"W"</span>, <span class="st">"Y"</span>, <span class="st">"F"</span>),
    <span class="kw">ST</span>  <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"S"</span>, <span class="st">"T"</span>),
    <span class="kw">CM</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"C"</span>, <span class="st">"M"</span>),
    <span class="kw">GALIV</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"G"</span>, <span class="st">"A"</span>, <span class="st">"L"</span>, <span class="st">"I"</span>, <span class="st">"V"</span>))
<span class="co">## set the colors by color blindness safe gradient color</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">colorBlindness</span>)
<span class="no">color</span> <span class="kw">&lt;-</span> <span class="no">Blue2Gray8Steps</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">color</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">group</span>)
<span class="co">## assign symbols for each group</span>
<span class="no">symbol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substring</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">group</span>), <span class="kw">first</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">last</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">symbol</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">group</span>)
<span class="co">## Add a grouping scheme</span>
<span class="co">## Please note that the user-supplied grouping is named </span>
<span class="co">## as "custom_group" internally.</span>
<span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/addScheme.html">addScheme</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">color</span>, <span class="kw">symbol</span> <span class="kw">=</span> <span class="no">symbol</span>, <span class="kw">group</span> <span class="kw">=</span> <span class="no">group</span>)
<span class="no">tt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/testDAU.html">testDAU</a></span>(<span class="kw">dagPeptides</span> <span class="kw">=</span> <span class="no">seq</span>, <span class="kw">dagBackground</span> <span class="kw">=</span> <span class="no">bg_ztest</span>,
              <span class="kw">groupingScheme</span> <span class="kw">=</span> <span class="st">"custom_group"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/dagLogo/man/dagLogo.html">dagLogo</a></span>(<span class="no">tt</span>, <span class="kw">legend</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<div class="figure">
<img src="motifStack_files/figure-html/customGroup-1.png" alt="Visualizing conserved amino acid sequence pattern based on chemical properties and hydrophobicity" width="576"><p class="caption">
Visualizing conserved amino acid sequence pattern based on chemical properties and hydrophobicity
</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-das2007survey">
<p>1. Das, M. K. &amp; Dai, H.-K. A survey of dna motif finding algorithms. in <em>BMC bioinformatics</em> <strong>8,</strong> S21 (BioMed Central, 2007).</p>
</div>
<div id="ref-sandve2007improved">
<p>2. Sandve, G. K., Abul, O., Walseng, V. &amp; Drabløs, F. Improved benchmarks for computational motif discovery. <em>BMC bioinformatics</em> <strong>8,</strong> 193 (2007).</p>
</div>
<div id="ref-tompa2005assessing">
<p>3. Tompa, M. <em>et al.</em> Assessing computational tools for the discovery of transcription factor binding sites. <em>Nature biotechnology</em> <strong>23,</strong> 137 (2005).</p>
</div>
<div id="ref-schneider1990sequence">
<p>4. Schneider, T. D. &amp; Stephens, R. M. Sequence logos: A new way to display consensus sequences. <em>Nucleic acids research</em> <strong>18,</strong> 6097–6100 (1990).</p>
</div>
<div id="ref-schneider1996reading">
<p>5. Schneider, T. D. in <em>Methods in enzymology</em> <strong>274,</strong> 445–455 (Elsevier, 1996).</p>
</div>
<div id="ref-schneider2001strong">
<p>6. Schneider, T. D. Strong minor groove base conservation in sequence logos implies dna distortion or base flipping during replication and transcription initiation. <em>Nucleic acids research</em> <strong>29,</strong> 4881–4891 (2001).</p>
</div>
<div id="ref-crooks2004weblogo">
<p>7. Crooks, G. E., Hon, G., Chandonia, J.-M. &amp; Brenner, S. E. WebLogo: A sequence logo generator. <em>Genome research</em> <strong>14,</strong> 1188–1190 (2004).</p>
</div>
<div id="ref-workman2005enologos">
<p>8. Workman, C. T. <em>et al.</em> EnoLOGOS: A versatile web tool for energy normalized sequence logos. <em>Nucleic acids research</em> <strong>33,</strong> W389–W392 (2005).</p>
</div>
<div id="ref-newburger2008uniprobe">
<p>9. Newburger, D. E. &amp; Bulyk, M. L. UniPROBE: An online database of protein binding microarray data on protein–dna interactions. <em>Nucleic acids research</em> <strong>37,</strong> D77–D82 (2008).</p>
</div>
<div id="ref-meng2005bacterial">
<p>10. Meng, X., Brodsky, M. H. &amp; Wolfe, S. A. A bacterial one-hybrid system for determining the dna-binding specificity of transcription factors. <em>Nature biotechnology</em> <strong>23,</strong> 988 (2005).</p>
</div>
<div id="ref-ellington1990vitro">
<p>11. Ellington, A. D. &amp; Szostak, J. W. In vitro selection of rna molecules that bind specific ligands. <em>nature</em> <strong>346,</strong> 818 (1990).</p>
</div>
<div id="ref-tuerk1990systematic">
<p>12. Tuerk, C. &amp; Gold, L. Systematic evolution of ligands by exponential enrichment: RNA ligands to bacteriophage t4 dna polymerase. <em>Science</em> <strong>249,</strong> 505–510 (1990).</p>
</div>
<div id="ref-hoinka2015large">
<p>13. Hoinka, J. <em>et al.</em> Large scale analysis of the mutational landscape in ht-selex improves aptamer discovery. <em>Nucleic acids research</em> <strong>43,</strong> 5699–5707 (2015).</p>
</div>
<div id="ref-isakova2017smile">
<p>14. Isakova, A. <em>et al.</em> SMiLE-seq identifies binding motifs of single and dimeric transcription factors. <em>Nature methods</em> <strong>14,</strong> 316 (2017).</p>
</div>
<div id="ref-bailey2009meme">
<p>15. Bailey, T. L. <em>et al.</em> MEME suite: Tools for motif discovery and searching. <em>Nucleic acids research</em> <strong>37,</strong> W202–W208 (2009).</p>
</div>
<div id="ref-badis2009diversity">
<p>16. Badis, G. <em>et al.</em> Diversity and complexity in dna recognition by transcription factors. <em>Science</em> <strong>324,</strong> 1720–1723 (2009).</p>
</div>
<div id="ref-weirauch2013evaluation">
<p>17. Weirauch, M. T. <em>et al.</em> Evaluation of methods for modeling transcription factor sequence specificity. <em>Nature biotechnology</em> <strong>31,</strong> 126 (2013).</p>
</div>
<div id="ref-zhao2011quantitative">
<p>18. Zhao, Y. &amp; Stormo, G. D. Quantitative analysis demonstrates most transcription factors require only simple models of specificity. <em>Nature biotechnology</em> <strong>29,</strong> 480 (2011).</p>
</div>
<div id="ref-bembom1seqlogo">
<p>19. Bembom, O. SeqLogo: Sequence logos for dna sequence alignments. <em>R package version</em> <strong>1,</strong></p>
</div>
<div id="ref-gupta2007quantifying">
<p>20. Gupta, S., Stamatoyannopoulos, J. A., Bailey, T. L. &amp; Noble, W. S. Quantifying similarity between motifs. <em>Genome biology</em> <strong>8,</strong> R24 (2007).</p>
</div>
<div id="ref-zhao2012conserved">
<p>21. Zhao, G. <em>et al.</em> Conserved motifs and prediction of regulatory modules in caenorhabditis elegans. <em>G3: Genes, Genomes, Genetics</em> <strong>2,</strong> 469–481 (2012).</p>
</div>
<div id="ref-mahony2007stamp">
<p>22. Mahony, S. &amp; Benos, P. V. STAMP: A web tool for exploring dna-binding motif similarities. <em>Nucleic acids research</em> <strong>35,</strong> W253–W258 (2007).</p>
</div>
<div id="ref-ou2018motifstack">
<p>23. Ou, J., Wolfe, S. A., Brodsky, M. H. &amp; Zhu, L. J. MotifStack for the analysis of transcription factor binding site evolution. <em>Nature methods</em> <strong>15,</strong> 8 (2018).</p>
</div>
<div id="ref-wingender2000transfac">
<p>24. Wingender, E. <em>et al.</em> TRANSFAC: An integrated system for gene expression regulation. <em>Nucleic acids research</em> <strong>28,</strong> 316–319 (2000).</p>
</div>
<div id="ref-sandelin2004jaspar">
<p>25. Sandelin, A., Alkema, W., Engström, P., Wasserman, W. W. &amp; Lenhard, B. JASPAR: An open-access database for eukaryotic transcription factor binding profiles. <em>Nucleic acids research</em> <strong>32,</strong> D91–D94 (2004).</p>
</div>
<div id="ref-weirauch2014determination">
<p>26. Weirauch, M. T. <em>et al.</em> Determination and inference of eukaryotic transcription factor sequence specificity. <em>Cell</em> <strong>158,</strong> 1431–1443 (2014).</p>
</div>
<div id="ref-kulakovskiy2017hocomoco">
<p>27. Kulakovskiy, I. V. <em>et al.</em> HOCOMOCO: Towards a complete collection of transcription factor binding models for human and mouse via large-scale chip-seq analysis. <em>Nucleic acids research</em> <strong>46,</strong> D252–D259 (2017).</p>
</div>
<div id="ref-zhu2010flyfactorsurvey">
<p>28. Zhu, L. J. <em>et al.</em> FlyFactorSurvey: A database of drosophila transcription factor binding specificities determined using the bacterial one-hybrid system. <em>Nucleic acids research</em> <strong>39,</strong> D111–D117 (2010).</p>
</div>
<div id="ref-heinz2010simple">
<p>29. Heinz, S. <em>et al.</em> Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and b cell identities. <em>Molecular cell</em> <strong>38,</strong> 576–589 (2010).</p>
</div>
<div id="ref-xie2009hpdi">
<p>30. Xie, Z., Hu, S., Blackshaw, S., Zhu, H. &amp; Qian, J. HPDI: A database of experimental human protein–dna interactions. <em>Bioinformatics</em> <strong>26,</strong> 287–289 (2009).</p>
</div>
<div id="ref-jolma2013dna">
<p>31. Jolma, A. <em>et al.</em> DNA-binding specificities of human transcription factors. <em>Cell</em> <strong>152,</strong> 327–339 (2013).</p>
</div>
<div id="ref-spivak2011scertf">
<p>32. Spivak, A. T. &amp; Stormo, G. D. ScerTF: A comprehensive database of benchmarked position weight matrices for saccharomyces species. <em>Nucleic acids research</em> <strong>40,</strong> D162–D168 (2011).</p>
</div>
<div id="ref-neph2012circuitry">
<p>33. Neph, S. <em>et al.</em> Circuitry and dynamics of human transcription factor regulatory networks. <em>Cell</em> <strong>150,</strong> 1274–1286 (2012).</p>
</div>
<div id="ref-pachkov2006swissregulon">
<p>34. Pachkov, M., Erb, I., Molina, N. &amp; Van Nimwegen, E. SwissRegulon: A database of genome-wide annotations of regulatory sites. <em>Nucleic acids research</em> <strong>35,</strong> D127–D131 (2006).</p>
</div>
<div id="ref-murrell2009importing">
<p>35. Murrell, P. &amp; others. Importing vector graphics: The grImport package for r. <em>Journal of Statistical Software</em> <strong>30,</strong> 1–37 (2009).</p>
</div>
<div id="ref-wickham2016ggplot2">
<p>36. Wickham, H. <em>Ggplot2: Elegant graphics for data analysis</em>. (Springer, 2016).</p>
</div>
<div id="ref-roepcke2005t">
<p>37. Roepcke, S., Grossmann, S., Rahmann, S. &amp; Vingron, M. T-reg comparator: An analysis tool for the comparison of position weight matrices. <em>Nucleic acids research</em> <strong>33,</strong> W438–W441 (2005).</p>
</div>
<div id="ref-dray2007ade4">
<p>38. Dray, S., Dufour, A.-B. &amp; others. The ade4 package: Implementing the duality diagram for ecologists. <em>Journal of statistical software</em> <strong>22,</strong> 1–20 (2007).</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
